{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "037e882b-90a3-47e8-9b4b-f560fa1b68d3",
   "metadata": {},
   "source": [
    "# Testing NCBI hg38.p14 GTF file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9342a73-3467-414b-89c8-d124fd8bf46f",
   "metadata": {},
   "source": [
    "*Anusha Aggarwal*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "2676b7da-c1d2-47c7-8c7f-682e0f455ec0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "tqdm.pandas()\n",
    "import re\n",
    "import os\n",
    "from pyfaidx import Fasta\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "60642a5d-23d4-47af-9843-e7c3adf4c458",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_cds = pd.read_csv('/mnt/home/aaggarwal/ceph/gates_proj/ncbi_genome_hg38.p14/hg38.p14.ncbiRefSeq.CDS_final_gtf.csv')\n",
    "gtf_trans = pd.read_csv('/mnt/home/aaggarwal/ceph/gates_proj/ncbi_genome_hg38.p14/hg38.p14.ncbiRefSeq.transcript_final_gtf.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "054a09b0-45ed-4013-aa7f-d85949f6a7f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>65565</td>\n",
       "      <td>65573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>79501</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>69037</td>\n",
       "      <td>70005</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>79501</td>\n",
       "      <td>969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>65565</td>\n",
       "      <td>65567</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>79501</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>70006</td>\n",
       "      <td>70008</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>79501</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>450743</td>\n",
       "      <td>451678</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"OR4F29\"; transcript_id \"NM_001005221....</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>NM_001005221.2</td>\n",
       "      <td>729759</td>\n",
       "      <td>936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229586</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>25038098</td>\n",
       "      <td>25038100</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"BPY2C\"; transcript_id \"NM_001002761.1...</td>\n",
       "      <td>BPY2C</td>\n",
       "      <td>NM_001002761.1</td>\n",
       "      <td>442868</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229587</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>25622443</td>\n",
       "      <td>25624034</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>NM_004680.3</td>\n",
       "      <td>9085</td>\n",
       "      <td>1592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229588</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>25624455</td>\n",
       "      <td>25624524</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>NM_004680.3</td>\n",
       "      <td>9085</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229589</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>25622443</td>\n",
       "      <td>25622445</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>NM_004680.3</td>\n",
       "      <td>9085</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229590</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>25624525</td>\n",
       "      <td>25624527</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>NM_004680.3</td>\n",
       "      <td>9085</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>229591 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname      source      feature     start       end score strand  \\\n",
       "0         chr1  BestRefSeq          CDS     65565     65573     .      +   \n",
       "1         chr1  BestRefSeq          CDS     69037     70005     .      +   \n",
       "2         chr1  BestRefSeq  start_codon     65565     65567     .      +   \n",
       "3         chr1  BestRefSeq   stop_codon     70006     70008     .      +   \n",
       "4         chr1  BestRefSeq          CDS    450743    451678     .      -   \n",
       "...        ...         ...          ...       ...       ...   ...    ...   \n",
       "229586    chrY  BestRefSeq   stop_codon  25038098  25038100     .      -   \n",
       "229587    chrY  BestRefSeq          CDS  25622443  25624034     .      +   \n",
       "229588    chrY  BestRefSeq          CDS  25624455  25624524     .      +   \n",
       "229589    chrY  BestRefSeq  start_codon  25622443  25622445     .      +   \n",
       "229590    chrY  BestRefSeq   stop_codon  25624525  25624527     .      +   \n",
       "\n",
       "        frame                                         attributes gene_id  \\\n",
       "0           0  gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...   OR4F5   \n",
       "1           0  gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...   OR4F5   \n",
       "2           0  gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...   OR4F5   \n",
       "3           0  gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...   OR4F5   \n",
       "4           0  gene_id \"OR4F29\"; transcript_id \"NM_001005221....  OR4F29   \n",
       "...       ...                                                ...     ...   \n",
       "229586      0  gene_id \"BPY2C\"; transcript_id \"NM_001002761.1...   BPY2C   \n",
       "229587      0  gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...    CDY1   \n",
       "229588      1  gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...    CDY1   \n",
       "229589      0  gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...    CDY1   \n",
       "229590      0  gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...    CDY1   \n",
       "\n",
       "         transcript_id  entrez_id  length  \n",
       "0       NM_001005484.2      79501       9  \n",
       "1       NM_001005484.2      79501     969  \n",
       "2       NM_001005484.2      79501       3  \n",
       "3       NM_001005484.2      79501       3  \n",
       "4       NM_001005221.2     729759     936  \n",
       "...                ...        ...     ...  \n",
       "229586  NM_001002761.1     442868       3  \n",
       "229587     NM_004680.3       9085    1592  \n",
       "229588     NM_004680.3       9085      70  \n",
       "229589     NM_004680.3       9085       3  \n",
       "229590     NM_004680.3       9085       3  \n",
       "\n",
       "[229591 rows x 13 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_cds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "b9b43131-49a8-47d9-b872-f2d532b60c2d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>entrez_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>65419</td>\n",
       "      <td>71585</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>79501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>450740</td>\n",
       "      <td>451678</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"OR4F29\"; transcript_id \"NM_001005221....</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>NM_001005221.2</td>\n",
       "      <td>729759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>685716</td>\n",
       "      <td>686654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"OR4F16\"; transcript_id \"NM_001005277....</td>\n",
       "      <td>OR4F16</td>\n",
       "      <td>NM_001005277.1</td>\n",
       "      <td>81399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>923923</td>\n",
       "      <td>944574</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"SAMD11\"; transcript_id \"NM_001385641....</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>NM_001385641.1</td>\n",
       "      <td>148398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>944203</td>\n",
       "      <td>959256</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658.4</td>\n",
       "      <td>26155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19161</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>24618004</td>\n",
       "      <td>24639207</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"BPY2B\"; transcript_id \"NM_001002760.1...</td>\n",
       "      <td>BPY2B</td>\n",
       "      <td>NM_001002760.1</td>\n",
       "      <td>442867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19162</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>24763069</td>\n",
       "      <td>24813393</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"DAZ3\"; transcript_id \"NM_020364.4\"; d...</td>\n",
       "      <td>DAZ3</td>\n",
       "      <td>NM_020364.4</td>\n",
       "      <td>57054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19163</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>24833919</td>\n",
       "      <td>24907040</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"DAZ4\"; transcript_id \"NM_001388484.1\"...</td>\n",
       "      <td>DAZ4</td>\n",
       "      <td>NM_001388484.1</td>\n",
       "      <td>57135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19164</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>25030901</td>\n",
       "      <td>25052104</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"BPY2C\"; transcript_id \"NM_001002761.1...</td>\n",
       "      <td>BPY2C</td>\n",
       "      <td>NM_001002761.1</td>\n",
       "      <td>442868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19165</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>25622117</td>\n",
       "      <td>25624902</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>NM_004680.3</td>\n",
       "      <td>9085</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19166 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqname      source     feature     start       end score strand frame  \\\n",
       "0        chr1  BestRefSeq  transcript     65419     71585     .      +     .   \n",
       "1        chr1  BestRefSeq  transcript    450740    451678     .      -     .   \n",
       "2        chr1  BestRefSeq  transcript    685716    686654     .      -     .   \n",
       "3        chr1  BestRefSeq  transcript    923923    944574     .      +     .   \n",
       "4        chr1  BestRefSeq  transcript    944203    959256     .      -     .   \n",
       "...       ...         ...         ...       ...       ...   ...    ...   ...   \n",
       "19161    chrY  BestRefSeq  transcript  24618004  24639207     .      +     .   \n",
       "19162    chrY  BestRefSeq  transcript  24763069  24813393     .      -     .   \n",
       "19163    chrY  BestRefSeq  transcript  24833919  24907040     .      +     .   \n",
       "19164    chrY  BestRefSeq  transcript  25030901  25052104     .      -     .   \n",
       "19165    chrY  BestRefSeq  transcript  25622117  25624902     .      +     .   \n",
       "\n",
       "                                              attributes gene_id  \\\n",
       "0      gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...   OR4F5   \n",
       "1      gene_id \"OR4F29\"; transcript_id \"NM_001005221....  OR4F29   \n",
       "2      gene_id \"OR4F16\"; transcript_id \"NM_001005277....  OR4F16   \n",
       "3      gene_id \"SAMD11\"; transcript_id \"NM_001385641....  SAMD11   \n",
       "4      gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...   NOC2L   \n",
       "...                                                  ...     ...   \n",
       "19161  gene_id \"BPY2B\"; transcript_id \"NM_001002760.1...   BPY2B   \n",
       "19162  gene_id \"DAZ3\"; transcript_id \"NM_020364.4\"; d...    DAZ3   \n",
       "19163  gene_id \"DAZ4\"; transcript_id \"NM_001388484.1\"...    DAZ4   \n",
       "19164  gene_id \"BPY2C\"; transcript_id \"NM_001002761.1...   BPY2C   \n",
       "19165  gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...    CDY1   \n",
       "\n",
       "        transcript_id  entrez_id  \n",
       "0      NM_001005484.2      79501  \n",
       "1      NM_001005221.2     729759  \n",
       "2      NM_001005277.1      81399  \n",
       "3      NM_001385641.1     148398  \n",
       "4         NM_015658.4      26155  \n",
       "...               ...        ...  \n",
       "19161  NM_001002760.1     442867  \n",
       "19162     NM_020364.4      57054  \n",
       "19163  NM_001388484.1      57135  \n",
       "19164  NM_001002761.1     442868  \n",
       "19165     NM_004680.3       9085  \n",
       "\n",
       "[19166 rows x 12 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_trans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "c656702f-d964-4557-840c-fd2ed3d01f63",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19166"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_gene_ids_count = gtf_cds[\"gene_id\"].nunique()\n",
    "unique_gene_ids_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "b731efea-13fd-4b25-aa43-6af02534212d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19166"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_gene_ids_count = gtf_cds[\"entrez_id\"].nunique()\n",
    "unique_gene_ids_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "id": "c649615d-802a-4e56-80b1-eeec52255d35",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_adjusted_cds_windows(transcript_id, chrom, strand, gtf_df):\n",
    "    transcript_df = gtf_df[gtf_df['transcript_id'] == transcript_id]\n",
    "    cds_df = transcript_df[transcript_df['feature'] == 'CDS']\n",
    "    start_df = transcript_df[transcript_df['feature'] == 'start_codon']\n",
    "    stop_df = transcript_df[transcript_df['feature'] == 'stop_codon']\n",
    "\n",
    "    if cds_df.empty or start_df.empty or stop_df.empty:\n",
    "        return None\n",
    "\n",
    "    start_codon_start = start_df.iloc[0]['start']\n",
    "    stop_codon_end = stop_df.iloc[0]['end']\n",
    "\n",
    "    if strand == '+':\n",
    "        cds_df = cds_df[\n",
    "            (cds_df['end'] >= start_codon_start) &\n",
    "            (cds_df['start'] <= stop_codon_end)\n",
    "        ].sort_values(by='start')\n",
    "    else:\n",
    "        start_codon_end = start_df.iloc[0]['end']\n",
    "        stop_codon_start = stop_df.iloc[0]['start']\n",
    "    \n",
    "        cds_df = cds_df[\n",
    "            (cds_df['end'] >= stop_codon_start) &\n",
    "            (cds_df['start'] <= start_codon_end)\n",
    "        ].sort_values(by='end', ascending=False)\n",
    "\n",
    "    if cds_df.empty:\n",
    "        return None\n",
    "\n",
    "    cds_windows = []\n",
    "    for i, row in cds_df.iterrows():\n",
    "        start = row['start']\n",
    "        end = row['end']\n",
    "\n",
    "        if strand == '+':\n",
    "            if start <= start_codon_start <= end:\n",
    "                start = start_codon_start\n",
    "        else:\n",
    "            if start <= start_codon_end <= end:\n",
    "                end = start_codon_end\n",
    "\n",
    "        if strand == '+':\n",
    "            if start <= stop_codon_end <= end:\n",
    "                end = stop_codon_end\n",
    "        else:\n",
    "            if start <= stop_codon_start <= end:\n",
    "                start = stop_codon_start\n",
    "\n",
    "        if end > start:\n",
    "            cds_windows.append((start, end, int(row['frame'])))\n",
    "\n",
    "    return cds_windows if cds_windows else None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "id": "ab5bdfe4-96d0-4ea1-b550-8cccbacd7886",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>539</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1919182</td>\n",
       "      <td>1919265</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...</td>\n",
       "      <td>TMEM52</td>\n",
       "      <td>NM_178545.4</td>\n",
       "      <td>339456</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>540</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1919045</td>\n",
       "      <td>1919088</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...</td>\n",
       "      <td>TMEM52</td>\n",
       "      <td>NM_178545.4</td>\n",
       "      <td>339456</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1918893</td>\n",
       "      <td>1918934</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...</td>\n",
       "      <td>TMEM52</td>\n",
       "      <td>NM_178545.4</td>\n",
       "      <td>339456</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>542</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1918253</td>\n",
       "      <td>1918431</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...</td>\n",
       "      <td>TMEM52</td>\n",
       "      <td>NM_178545.4</td>\n",
       "      <td>339456</td>\n",
       "      <td>179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1917885</td>\n",
       "      <td>1918162</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...</td>\n",
       "      <td>TMEM52</td>\n",
       "      <td>NM_178545.4</td>\n",
       "      <td>339456</td>\n",
       "      <td>278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>544</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1919263</td>\n",
       "      <td>1919265</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...</td>\n",
       "      <td>TMEM52</td>\n",
       "      <td>NM_178545.4</td>\n",
       "      <td>339456</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>545</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>1917882</td>\n",
       "      <td>1917884</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...</td>\n",
       "      <td>TMEM52</td>\n",
       "      <td>NM_178545.4</td>\n",
       "      <td>339456</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqname      source      feature    start      end score strand  frame  \\\n",
       "539    chr1  BestRefSeq          CDS  1919182  1919265     .      -      0   \n",
       "540    chr1  BestRefSeq          CDS  1919045  1919088     .      -      0   \n",
       "541    chr1  BestRefSeq          CDS  1918893  1918934     .      -      1   \n",
       "542    chr1  BestRefSeq          CDS  1918253  1918431     .      -      1   \n",
       "543    chr1  BestRefSeq          CDS  1917885  1918162     .      -      2   \n",
       "544    chr1  BestRefSeq  start_codon  1919263  1919265     .      -      0   \n",
       "545    chr1  BestRefSeq   stop_codon  1917882  1917884     .      -      0   \n",
       "\n",
       "                                            attributes gene_id transcript_id  \\\n",
       "539  gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...  TMEM52   NM_178545.4   \n",
       "540  gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...  TMEM52   NM_178545.4   \n",
       "541  gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...  TMEM52   NM_178545.4   \n",
       "542  gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...  TMEM52   NM_178545.4   \n",
       "543  gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...  TMEM52   NM_178545.4   \n",
       "544  gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...  TMEM52   NM_178545.4   \n",
       "545  gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...  TMEM52   NM_178545.4   \n",
       "\n",
       "     entrez_id  length  \n",
       "539     339456      84  \n",
       "540     339456      44  \n",
       "541     339456      42  \n",
       "542     339456     179  \n",
       "543     339456     278  \n",
       "544     339456       3  \n",
       "545     339456       3  "
      ]
     },
     "execution_count": 232,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_cds[gtf_cds['entrez_id'] == 339456]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "id": "fd0e5e55-0d7c-4248-ac01-3ceeae8bb3e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>415</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1574478</td>\n",
       "      <td>1574557</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...</td>\n",
       "      <td>SSU72</td>\n",
       "      <td>NM_014188.3</td>\n",
       "      <td>29101</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>416</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1564773</td>\n",
       "      <td>1564916</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...</td>\n",
       "      <td>SSU72</td>\n",
       "      <td>NM_014188.3</td>\n",
       "      <td>29101</td>\n",
       "      <td>144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>417</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1544863</td>\n",
       "      <td>1545002</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...</td>\n",
       "      <td>SSU72</td>\n",
       "      <td>NM_014188.3</td>\n",
       "      <td>29101</td>\n",
       "      <td>140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>418</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1543869</td>\n",
       "      <td>1543987</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...</td>\n",
       "      <td>SSU72</td>\n",
       "      <td>NM_014188.3</td>\n",
       "      <td>29101</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1542069</td>\n",
       "      <td>1542167</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...</td>\n",
       "      <td>SSU72</td>\n",
       "      <td>NM_014188.3</td>\n",
       "      <td>29101</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>420</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1574555</td>\n",
       "      <td>1574557</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...</td>\n",
       "      <td>SSU72</td>\n",
       "      <td>NM_014188.3</td>\n",
       "      <td>29101</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>1542066</td>\n",
       "      <td>1542068</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...</td>\n",
       "      <td>SSU72</td>\n",
       "      <td>NM_014188.3</td>\n",
       "      <td>29101</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqname      source      feature    start      end score strand  frame  \\\n",
       "415    chr1  BestRefSeq          CDS  1574478  1574557     .      -      0   \n",
       "416    chr1  BestRefSeq          CDS  1564773  1564916     .      -      1   \n",
       "417    chr1  BestRefSeq          CDS  1544863  1545002     .      -      1   \n",
       "418    chr1  BestRefSeq          CDS  1543869  1543987     .      -      2   \n",
       "419    chr1  BestRefSeq          CDS  1542069  1542167     .      -      0   \n",
       "420    chr1  BestRefSeq  start_codon  1574555  1574557     .      -      0   \n",
       "421    chr1  BestRefSeq   stop_codon  1542066  1542068     .      -      0   \n",
       "\n",
       "                                            attributes gene_id transcript_id  \\\n",
       "415  gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...   SSU72   NM_014188.3   \n",
       "416  gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...   SSU72   NM_014188.3   \n",
       "417  gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...   SSU72   NM_014188.3   \n",
       "418  gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...   SSU72   NM_014188.3   \n",
       "419  gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...   SSU72   NM_014188.3   \n",
       "420  gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...   SSU72   NM_014188.3   \n",
       "421  gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...   SSU72   NM_014188.3   \n",
       "\n",
       "     entrez_id  length  \n",
       "415      29101      80  \n",
       "416      29101     144  \n",
       "417      29101     140  \n",
       "418      29101     119  \n",
       "419      29101      99  \n",
       "420      29101       3  \n",
       "421      29101       3  "
      ]
     },
     "execution_count": 216,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_cds[gtf_cds['transcript_id'] == 'NM_014188.3']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "id": "68ea8cc0-acf8-4d48-a205-5f76fb10f48d",
   "metadata": {},
   "outputs": [],
   "source": [
    "windows = get_adjusted_cds_windows(\"NM_014188.3\", \"chr1\", \"-\", gtf_cds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "id": "b7f90260-5489-42c3-a76c-0e2bed785b84",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(1574478, 1574557, 0),\n",
       " (1564773, 1564916, 1),\n",
       " (1544863, 1545002, 1),\n",
       " (1543869, 1543987, 2),\n",
       " (1542069, 1542167, 0)]"
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "windows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "id": "e9a4772a-8cbf-4fb3-b7f6-0b79cdf1c080",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_fasta_dir = \"/mnt/home/aaggarwal/ceph/gates_proj/ncbi_genome_hg38.p14\"\n",
    "\n",
    "# load the genome\n",
    "fasta_files = {\n",
    "    f.replace(\".fa\", \"\"): os.path.join(genome_fasta_dir, f)\n",
    "    for f in os.listdir(genome_fasta_dir) if f.endswith(\".fa\")\n",
    "}\n",
    "fasta_index = {k: Fasta(v) for k, v in fasta_files.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "id": "82d1e195-1285-4573-a2e7-c5457b278bfd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(1574478, 1574557, 0),\n",
       " (1564773, 1564916, 1),\n",
       " (1544863, 1545002, 1),\n",
       " (1543869, 1543987, 2),\n",
       " (1542069, 1542167, 0)]"
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds_windows = sorted(windows, key=lambda x: x[0], reverse=True)\n",
    "cds_windows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "id": "5c16ac2f-3ab8-426d-a4c9-d049b5598c2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = fasta_index['chr1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "id": "9ff18a50-3f50-4ef6-a428-bba536a36e54",
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom = genome['chr1'] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "id": "715138e5-7c0f-4a93-bd27-a8942d88fa87",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_parts = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "id": "8ce9fbf2-7ff8-4622-8f42-61cf28537ab6",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, (start, end, frame) in enumerate(cds_windows):\n",
    "    start = int(start)\n",
    "    end = int(end)\n",
    "    frame = int(frame)\n",
    "\n",
    "    exon_seq = chrom[start-1:end].seq.upper()\n",
    "\n",
    "    if i == 0:\n",
    "        exon_seq = exon_seq[frame:]\n",
    "\n",
    "    seq_parts.append(exon_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "id": "8a1a61b0-7aa0-46fa-b140-6470f56aa67d",
   "metadata": {},
   "outputs": [],
   "source": [
    "full_seq = ''.join(seq_parts[::-1])  # reverse exon order\n",
    "full_seq_rc = str(Seq(full_seq).reverse_complement())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "id": "27454f8a-8551-4779-96c7-56fe6d5c8767",
   "metadata": {},
   "outputs": [],
   "source": [
    "protein = str(Seq(full_seq_rc).translate(to_stop=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "id": "94fd8bb5-d358-422e-9023-cccb0b177cd7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First codon: ATG\n",
      "Protein start: MPSSPLRVAV\n",
      "Full protein: MPSSPLRVAVVCSSNQNRSMEAHNILSKRGFSVRSFGTGTHVKLPGPAPDKPNVYDFKTTYDQMYNDLLRKDKELYTQNGILHMLDRNKRIKPRPERFQNCKDLFDLILTCEERVYDQVVEDLNSREQETCQPVHVVNVDIQDNHEEATLGAFLICELCQCIQHTEDMENEIDELLQEFEEKSGRTFLHTVCFY\n"
     ]
    }
   ],
   "source": [
    "print(f\"First codon: {full_seq_rc[:3]}\")\n",
    "print(f\"Protein start: {protein[:10]}\")\n",
    "print(f\"Full protein: {protein}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "id": "6e1e7911-2472-4241-93c3-a1554b8f1f0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_parts = []\n",
    "for i, (start, end, frame) in enumerate(cds_windows):\n",
    "    start = int(start)\n",
    "    end = int(end)\n",
    "    frame = int(frame)\n",
    "\n",
    "    exon_seq = chrom[start-1:end].seq.upper()\n",
    "\n",
    "    exon_seq = exon_seq[frame:]\n",
    "\n",
    "    seq_parts.append(exon_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "id": "3017411f-9c13-447e-ad9e-6e86940f3331",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First codon: ATG\n",
      "Protein start: MPSSPLRVAV\n",
      "Full protein: MPSSPLRVAVVCSSNQNRSMEAHNILSKRGFSVRSFGTGTHVKLPGPAPDKPNVYDFKTTYDQMYNDLLRKDKEPIHRMGFYICWTEIRESSPGQKDSRTAKTCLI*SSLAKRECMTRWWKSEFQRTGDLPACARGQCGHPGQPRGGHPGGVSHL*ALPVSSTRKTWRTRSTSCCRSSRRRVAAPFCTPSAS\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/home/aaggarwal/miniconda3/envs/target-genes/lib/python3.10/site-packages/Bio/Seq.py:2879: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "full_seq = ''.join(seq_parts[::-1])  # reverse exon order\n",
    "full_seq_rc = str(Seq(full_seq).reverse_complement())\n",
    "\n",
    "protein = str(Seq(full_seq_rc).translate(to_stop=False))\n",
    "\n",
    "print(f\"First codon: {full_seq_rc[:3]}\")\n",
    "print(f\"Protein start: {protein[:10]}\")\n",
    "print(f\"Full protein: {protein}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "target-genes",
   "language": "python",
   "name": "target-genes"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
