{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "037e882b-90a3-47e8-9b4b-f560fa1b68d3",
   "metadata": {},
   "source": [
    "# Testing NCBI hg38.p14 GTF file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9342a73-3467-414b-89c8-d124fd8bf46f",
   "metadata": {},
   "source": [
    "*Anusha Aggarwal*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "2676b7da-c1d2-47c7-8c7f-682e0f455ec0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "45559543-bc2b-4b49-8f2a-00728c21c4bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>entrez_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>65565</td>\n",
       "      <td>65573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>79501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>69037</td>\n",
       "      <td>70005</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>79501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>65565</td>\n",
       "      <td>65567</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>79501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>70006</td>\n",
       "      <td>70008</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>79501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>450743</td>\n",
       "      <td>451678</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"OR4F29\"; transcript_id \"NM_001005221....</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>NM_001005221.2</td>\n",
       "      <td>729759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230230</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>25038098</td>\n",
       "      <td>25038100</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"BPY2C\"; transcript_id \"NM_001002761.1...</td>\n",
       "      <td>BPY2C</td>\n",
       "      <td>NM_001002761.1</td>\n",
       "      <td>442868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230231</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>25622443</td>\n",
       "      <td>25624034</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>NM_004680.3</td>\n",
       "      <td>9085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230232</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>25624455</td>\n",
       "      <td>25624524</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>NM_004680.3</td>\n",
       "      <td>9085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230233</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>25622443</td>\n",
       "      <td>25622445</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>NM_004680.3</td>\n",
       "      <td>9085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230234</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>25624525</td>\n",
       "      <td>25624527</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>NM_004680.3</td>\n",
       "      <td>9085</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>230235 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname      source      feature     start       end score strand  \\\n",
       "0         chr1  BestRefSeq          CDS     65565     65573     .      +   \n",
       "1         chr1  BestRefSeq          CDS     69037     70005     .      +   \n",
       "2         chr1  BestRefSeq  start_codon     65565     65567     .      +   \n",
       "3         chr1  BestRefSeq   stop_codon     70006     70008     .      +   \n",
       "4         chr1  BestRefSeq          CDS    450743    451678     .      -   \n",
       "...        ...         ...          ...       ...       ...   ...    ...   \n",
       "230230    chrY  BestRefSeq   stop_codon  25038098  25038100     .      -   \n",
       "230231    chrY  BestRefSeq          CDS  25622443  25624034     .      +   \n",
       "230232    chrY  BestRefSeq          CDS  25624455  25624524     .      +   \n",
       "230233    chrY  BestRefSeq  start_codon  25622443  25622445     .      +   \n",
       "230234    chrY  BestRefSeq   stop_codon  25624525  25624527     .      +   \n",
       "\n",
       "       frame                                         attributes gene_id  \\\n",
       "0          0  gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...   OR4F5   \n",
       "1          0  gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...   OR4F5   \n",
       "2          0  gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...   OR4F5   \n",
       "3          0  gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...   OR4F5   \n",
       "4          0  gene_id \"OR4F29\"; transcript_id \"NM_001005221....  OR4F29   \n",
       "...      ...                                                ...     ...   \n",
       "230230     0  gene_id \"BPY2C\"; transcript_id \"NM_001002761.1...   BPY2C   \n",
       "230231     0  gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...    CDY1   \n",
       "230232     1  gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...    CDY1   \n",
       "230233     0  gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...    CDY1   \n",
       "230234     0  gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...    CDY1   \n",
       "\n",
       "         transcript_id entrez_id  \n",
       "0       NM_001005484.2     79501  \n",
       "1       NM_001005484.2     79501  \n",
       "2       NM_001005484.2     79501  \n",
       "3       NM_001005484.2     79501  \n",
       "4       NM_001005221.2    729759  \n",
       "...                ...       ...  \n",
       "230230  NM_001002761.1    442868  \n",
       "230231     NM_004680.3      9085  \n",
       "230232     NM_004680.3      9085  \n",
       "230233     NM_004680.3      9085  \n",
       "230234     NM_004680.3      9085  \n",
       "\n",
       "[230235 rows x 12 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_path = '/mnt/home/aaggarwal/ceph/gates_proj/ncbi_genome_hg38.p14/GCF_000001405.40_GRCh38.p14_genomic.final_CDS.gtf'\n",
    "\n",
    "col_names = [\n",
    "    \"seqname\", \"source\", \"feature\", \"start\", \"end\",\n",
    "    \"score\", \"strand\", \"frame\", \"attributes\"\n",
    "]\n",
    "\n",
    "gtf = pd.read_csv(\n",
    "    gtf_path,\n",
    "    sep='\\t',\n",
    "    comment='#',\n",
    "    names=col_names,\n",
    "    dtype=str\n",
    ")\n",
    "\n",
    "gtf['start'] = gtf['start'].astype(int)\n",
    "gtf['end'] = gtf['end'].astype(int)\n",
    "\n",
    "gtf['gene_id'] = gtf['attributes'].str.extract(r'gene_id \"([^\"]+)\"')\n",
    "gtf['transcript_id'] = gtf['attributes'].str.extract(r'transcript_id \"([^\"]+)\"')\n",
    "gtf['entrez_id'] = gtf['attributes'].str.extract(r'GeneID:(\\d+)')\n",
    "\n",
    "gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "ba2805ad-4297-4a1d-a512-49f57d004205",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf[\"length\"] = gtf[\"end\"] - gtf[\"start\"] + 1\n",
    "\n",
    "gtf[\"is_multiple_of_3\"] = gtf[\"length\"] % 3 == 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "1cae0c15-b958-470d-a6cc-6b43cbd34389",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>length</th>\n",
       "      <th>is_multiple_of_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>65565</td>\n",
       "      <td>65573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>79501</td>\n",
       "      <td>9</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>69037</td>\n",
       "      <td>70005</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>79501</td>\n",
       "      <td>969</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>65565</td>\n",
       "      <td>65567</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>79501</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>70006</td>\n",
       "      <td>70008</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>79501</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>450743</td>\n",
       "      <td>451678</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"OR4F29\"; transcript_id \"NM_001005221....</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>NM_001005221.2</td>\n",
       "      <td>729759</td>\n",
       "      <td>936</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230230</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>25038098</td>\n",
       "      <td>25038100</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"BPY2C\"; transcript_id \"NM_001002761.1...</td>\n",
       "      <td>BPY2C</td>\n",
       "      <td>NM_001002761.1</td>\n",
       "      <td>442868</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230231</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>25622443</td>\n",
       "      <td>25624034</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>NM_004680.3</td>\n",
       "      <td>9085</td>\n",
       "      <td>1592</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230232</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>25624455</td>\n",
       "      <td>25624524</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>NM_004680.3</td>\n",
       "      <td>9085</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230233</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>25622443</td>\n",
       "      <td>25622445</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>NM_004680.3</td>\n",
       "      <td>9085</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230234</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>25624525</td>\n",
       "      <td>25624527</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>NM_004680.3</td>\n",
       "      <td>9085</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>230235 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname      source      feature     start       end score strand  \\\n",
       "0         chr1  BestRefSeq          CDS     65565     65573     .      +   \n",
       "1         chr1  BestRefSeq          CDS     69037     70005     .      +   \n",
       "2         chr1  BestRefSeq  start_codon     65565     65567     .      +   \n",
       "3         chr1  BestRefSeq   stop_codon     70006     70008     .      +   \n",
       "4         chr1  BestRefSeq          CDS    450743    451678     .      -   \n",
       "...        ...         ...          ...       ...       ...   ...    ...   \n",
       "230230    chrY  BestRefSeq   stop_codon  25038098  25038100     .      -   \n",
       "230231    chrY  BestRefSeq          CDS  25622443  25624034     .      +   \n",
       "230232    chrY  BestRefSeq          CDS  25624455  25624524     .      +   \n",
       "230233    chrY  BestRefSeq  start_codon  25622443  25622445     .      +   \n",
       "230234    chrY  BestRefSeq   stop_codon  25624525  25624527     .      +   \n",
       "\n",
       "       frame                                         attributes gene_id  \\\n",
       "0          0  gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...   OR4F5   \n",
       "1          0  gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...   OR4F5   \n",
       "2          0  gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...   OR4F5   \n",
       "3          0  gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...   OR4F5   \n",
       "4          0  gene_id \"OR4F29\"; transcript_id \"NM_001005221....  OR4F29   \n",
       "...      ...                                                ...     ...   \n",
       "230230     0  gene_id \"BPY2C\"; transcript_id \"NM_001002761.1...   BPY2C   \n",
       "230231     0  gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...    CDY1   \n",
       "230232     1  gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...    CDY1   \n",
       "230233     0  gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...    CDY1   \n",
       "230234     0  gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...    CDY1   \n",
       "\n",
       "         transcript_id entrez_id  length  is_multiple_of_3  \n",
       "0       NM_001005484.2     79501       9              True  \n",
       "1       NM_001005484.2     79501     969              True  \n",
       "2       NM_001005484.2     79501       3              True  \n",
       "3       NM_001005484.2     79501       3              True  \n",
       "4       NM_001005221.2    729759     936              True  \n",
       "...                ...       ...     ...               ...  \n",
       "230230  NM_001002761.1    442868       3              True  \n",
       "230231     NM_004680.3      9085    1592             False  \n",
       "230232     NM_004680.3      9085      70             False  \n",
       "230233     NM_004680.3      9085       3              True  \n",
       "230234     NM_004680.3      9085       3              True  \n",
       "\n",
       "[230235 rows x 14 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "443d6778-0f26-4b95-94db-d5cc6901c6e6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "a0d4eb53-ee84-4c34-b325-761a983caf60",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "is_multiple_of_3\n",
      "True     118609\n",
      "False    111626\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(gtf['is_multiple_of_3'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "443f5d3a-7bad-4cd3-a5f0-b1237814cc25",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "is_multiple_of_3  False  True \n",
      "frame                         \n",
      "0                 45337  91248\n",
      "1                 29780   8025\n",
      "2                 36509  19336\n"
     ]
    }
   ],
   "source": [
    "breakdown = gtf.groupby(['frame', 'is_multiple_of_3']).size().unstack(fill_value=0)\n",
    "print(breakdown)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "ca727d85-82c1-4e94-a867-0d2b4939bfb9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>is_multiple_of_3</th>\n",
       "      <th>False</th>\n",
       "      <th>True</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>frame</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>45337</td>\n",
       "      <td>91248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>18175</td>\n",
       "      <td>19630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>30139</td>\n",
       "      <td>25706</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "is_multiple_of_3  False  True \n",
       "frame                         \n",
       "0                 45337  91248\n",
       "1                 18175  19630\n",
       "2                 30139  25706"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ensure correct types\n",
    "gtf[\"start\"] = gtf[\"start\"].astype(int)\n",
    "gtf[\"end\"] = gtf[\"end\"].astype(int)\n",
    "gtf[\"frame\"] = gtf[\"frame\"].fillna(\"0\").astype(str)\n",
    "\n",
    "# Adjust length based on frame\n",
    "def adjust_length(row):\n",
    "    frame = int(row[\"frame\"]) if row[\"frame\"].isdigit() else 0\n",
    "    length = row[\"end\"] - row[\"start\"] + 1\n",
    "    return length - frame\n",
    "\n",
    "gtf[\"adjusted_length\"] = gtf.apply(adjust_length, axis=1)\n",
    "gtf[\"is_multiple_of_3\"] = gtf[\"adjusted_length\"] % 3 == 0\n",
    "\n",
    "# Generate new stats\n",
    "adjusted_stats = gtf.groupby([\"frame\", \"is_multiple_of_3\"]).size().unstack(fill_value=0)\n",
    "adjusted_stats\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "fa65b976-346b-400f-a6c8-c10ea7e503fe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start codon matches CDS start in 50.69% of cases\n",
      "Stop codon matches CDS end in 0.00% of cases\n"
     ]
    }
   ],
   "source": [
    "# Subset GTF\n",
    "cds_df = gtf[gtf['feature'] == 'CDS'].copy()\n",
    "start_df = gtf[gtf['feature'] == 'start_codon'].copy()\n",
    "stop_df = gtf[gtf['feature'] == 'stop_codon'].copy()\n",
    "\n",
    "# Convert to numeric\n",
    "for col in ['start', 'end']:\n",
    "    cds_df[col] = pd.to_numeric(cds_df[col])\n",
    "    start_df[col] = pd.to_numeric(start_df[col])\n",
    "    stop_df[col] = pd.to_numeric(stop_df[col])\n",
    "\n",
    "# Group CDS per transcript\n",
    "cds_grp = cds_df.groupby('entrez_id').agg({\n",
    "    'start': ['min', 'max'],\n",
    "    'end': ['min', 'max'],\n",
    "    'strand': 'first'\n",
    "}).reset_index()\n",
    "\n",
    "cds_grp.columns = ['entrez_id', 'cds_start_min', 'cds_start_max', 'cds_end_min', 'cds_end_max', 'strand']\n",
    "\n",
    "# Get start_codon and stop_codon positions\n",
    "start_codon_pos = start_df.groupby('entrez_id')['start'].first().rename('start_codon_start')\n",
    "stop_codon_pos = stop_df.groupby('entrez_id')['end'].first().rename('stop_codon_end')\n",
    "\n",
    "# Merge all\n",
    "merged = cds_grp.merge(start_codon_pos, on='entrez_id', how='left')\n",
    "merged = merged.merge(stop_codon_pos, on='entrez_id', how='left')\n",
    "\n",
    "# Define check conditions\n",
    "def check_start(row):\n",
    "    if row['strand'] == '+':\n",
    "        return row['start_codon_start'] == row['cds_start_min']\n",
    "    else:\n",
    "        return row['start_codon_start'] == row['cds_end_max']\n",
    "\n",
    "def check_stop(row):\n",
    "    if row['strand'] == '+':\n",
    "        return row['stop_codon_end'] == row['cds_end_max']\n",
    "    else:\n",
    "        return row['stop_codon_end'] == row['cds_start_min']\n",
    "\n",
    "merged['start_matches'] = merged.apply(check_start, axis=1)\n",
    "merged['stop_matches'] = merged.apply(check_stop, axis=1)\n",
    "\n",
    "# Summary stats\n",
    "start_match_pct = merged['start_matches'].mean() * 100\n",
    "stop_match_pct = merged['stop_matches'].mean() * 100\n",
    "\n",
    "print(f\"Start codon matches CDS start in {start_match_pct:.2f}% of cases\")\n",
    "print(f\"Stop codon matches CDS end in {stop_match_pct:.2f}% of cases\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "c158d081-6c69-4998-85b5-c70266d808d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "frame\n",
       "0    38438\n",
       "1       42\n",
       "2       11\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter for start and stop codons\n",
    "codon_df = gtf[gtf['feature'].isin(['start_codon', 'stop_codon'])]\n",
    "\n",
    "# Check if all start/stop codons have frame == 0\n",
    "codon_frame_counts = codon_df['frame'].value_counts(dropna=False)\n",
    "\n",
    "codon_frame_counts\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "02452436-2c71-48b3-adef-2899dbf4f025",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Filter for start_codon and stop_codon features\n",
    "start_stop_codons = gtf[gtf[\"feature\"].isin([\"start_codon\", \"stop_codon\"])].copy()\n",
    "\n",
    "# Convert start and end to integers\n",
    "start_stop_codons[\"start\"] = start_stop_codons[\"start\"].astype(int)\n",
    "start_stop_codons[\"end\"] = start_stop_codons[\"end\"].astype(int)\n",
    "\n",
    "# Calculate lengths\n",
    "start_stop_codons[\"length\"] = start_stop_codons[\"end\"] - start_stop_codons[\"start\"] + 1\n",
    "\n",
    "# Group by feature and summarize length distribution\n",
    "length_summary = start_stop_codons.groupby(\"feature\")[\"length\"].value_counts().unstack().fillna(0).astype(int)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "c63f9ee5-4820-4c02-b29c-8bb3677ca024",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>length</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>start_codon</th>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>19186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stop_codon</th>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>19199</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "length        1   2      3\n",
       "feature                   \n",
       "start_codon  33  33  19186\n",
       "stop_codon   20  20  19199"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "length_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "297ff5a1-71aa-4151-9236-e6501a8f1d96",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Stop codon follows CDS end correctly in 99.82% of transcripts\n"
     ]
    }
   ],
   "source": [
    "# Ensure all numeric columns are proper integers\n",
    "gtf[[\"start\", \"end\"]] = gtf[[\"start\", \"end\"]].astype(int)\n",
    "\n",
    "# Split into feature types\n",
    "cds_df = gtf[gtf[\"feature\"] == \"CDS\"]\n",
    "stop_df = gtf[gtf[\"feature\"] == \"stop_codon\"]\n",
    "\n",
    "# Get last and first CDS entries per transcript\n",
    "last_cds = cds_df.sort_values(by=[\"transcript_id\", \"end\"]).groupby(\"transcript_id\").last().reset_index()\n",
    "first_cds = cds_df.sort_values(by=[\"transcript_id\", \"start\"]).groupby(\"transcript_id\").first().reset_index()\n",
    "\n",
    "# Keep necessary columns only\n",
    "last_cds = last_cds[[\"transcript_id\", \"end\"]].rename(columns={\"end\": \"cds_end\"})\n",
    "first_cds = first_cds[[\"transcript_id\", \"start\"]].rename(columns={\"start\": \"cds_start\"})\n",
    "\n",
    "# Prepare stop codons with transcript and strand\n",
    "stop_check = stop_df[[\"transcript_id\", \"start\", \"end\", \"strand\"]].copy()\n",
    "\n",
    "# Merge\n",
    "stop_check = stop_check.merge(last_cds, on=\"transcript_id\", how=\"left\")\n",
    "stop_check = stop_check.merge(first_cds, on=\"transcript_id\", how=\"left\")\n",
    "\n",
    "# Check for match, considering strand\n",
    "stop_check[\"stop_matches\"] = (\n",
    "    ((stop_check[\"strand\"] == \"+\") & (stop_check[\"start\"] == stop_check[\"cds_end\"] + 1)) |\n",
    "    ((stop_check[\"strand\"] == \"-\") & (stop_check[\"end\"] == stop_check[\"cds_start\"] - 1))\n",
    ")\n",
    "\n",
    "# Result\n",
    "print(f\"Stop codon follows CDS end correctly in {stop_check['stop_matches'].mean():.2%} of transcripts\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "b15a2a06-0137-434d-9a0f-a6f8ba1e840f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of entries with MANE Select: 230235\n"
     ]
    }
   ],
   "source": [
    "mane_select_count = gtf['attributes'].str.contains('MANE Select', na=False).sum()\n",
    "print(\"Number of entries with MANE Select:\", mane_select_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "56584db4-f845-459e-a234-32b66273168a",
   "metadata": {},
   "outputs": [],
   "source": [
    "transcript_counts = gtf.groupby(\"gene_id\")[\"transcript_id\"].nunique()\n",
    "\n",
    "multiple_transcripts = transcript_counts[transcript_counts > 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "c98cbd31-bfe1-46bc-ac68-d9d60efb050c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: transcript_id, dtype: int64)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multiple_transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "77d896ac-b8eb-41fa-a9ec-1bd0487a7eea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19219"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_gene_ids_count = gtf[\"gene_id\"].nunique()\n",
    "unique_gene_ids_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "48babfeb-4af6-499b-b007-9762b621c972",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19219"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_gene_ids_count = gtf[\"entrez_id\"].nunique()\n",
    "unique_gene_ids_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "1b4ec670-de8b-49dd-a795-e7012686deaf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19219"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_gene_ids_count = gtf[\"transcript_id\"].nunique()\n",
    "unique_gene_ids_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "f4749d77-1a2c-410e-9aea-ef76dca1551c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes with multiple MANE Select transcripts:\n",
      "Series([], Name: transcript_id, dtype: int64)\n",
      "Genes with no MANE Select transcript:\n",
      "set()\n"
     ]
    }
   ],
   "source": [
    "mane_df = gtf[gtf['attributes'].str.contains('tag \"MANE Select\"')]\n",
    "\n",
    "gene_mane_counts = mane_df.groupby('gene_id')['transcript_id'].nunique()\n",
    "\n",
    "multiple_mane = gene_mane_counts[gene_mane_counts > 1]\n",
    "print(\"Genes with multiple MANE Select transcripts:\")\n",
    "print(multiple_mane)\n",
    "\n",
    "all_gene_ids = gtf['gene_id'].unique()\n",
    "mane_gene_ids = mane_df['gene_id'].unique()\n",
    "missing_mane = set(all_gene_ids) - set(mane_gene_ids)\n",
    "print(\"Genes with no MANE Select transcript:\")\n",
    "print(missing_mane)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "10a1f864-471d-4007-9174-ab2bb1f6395d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>entrez_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>65419</td>\n",
       "      <td>71585</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>79501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>450740</td>\n",
       "      <td>451678</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"OR4F29\"; transcript_id \"NM_001005221....</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>NM_001005221.2</td>\n",
       "      <td>729759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>685716</td>\n",
       "      <td>686654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"OR4F16\"; transcript_id \"NM_001005277....</td>\n",
       "      <td>OR4F16</td>\n",
       "      <td>NM_001005277.1</td>\n",
       "      <td>81399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>923923</td>\n",
       "      <td>944574</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"SAMD11\"; transcript_id \"NM_001385641....</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>NM_001385641.1</td>\n",
       "      <td>148398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>944203</td>\n",
       "      <td>959256</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658.4</td>\n",
       "      <td>26155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19214</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>24618004</td>\n",
       "      <td>24639207</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"BPY2B\"; transcript_id \"NM_001002760.1...</td>\n",
       "      <td>BPY2B</td>\n",
       "      <td>NM_001002760.1</td>\n",
       "      <td>442867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19215</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>24763069</td>\n",
       "      <td>24813393</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"DAZ3\"; transcript_id \"NM_020364.4\"; d...</td>\n",
       "      <td>DAZ3</td>\n",
       "      <td>NM_020364.4</td>\n",
       "      <td>57054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19216</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>24833919</td>\n",
       "      <td>24907040</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"DAZ4\"; transcript_id \"NM_001388484.1\"...</td>\n",
       "      <td>DAZ4</td>\n",
       "      <td>NM_001388484.1</td>\n",
       "      <td>57135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19217</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>25030901</td>\n",
       "      <td>25052104</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"BPY2C\"; transcript_id \"NM_001002761.1...</td>\n",
       "      <td>BPY2C</td>\n",
       "      <td>NM_001002761.1</td>\n",
       "      <td>442868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19218</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>25622117</td>\n",
       "      <td>25624902</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>NM_004680.3</td>\n",
       "      <td>9085</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19219 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqname      source     feature     start       end score strand frame  \\\n",
       "0        chr1  BestRefSeq  transcript     65419     71585     .      +     .   \n",
       "1        chr1  BestRefSeq  transcript    450740    451678     .      -     .   \n",
       "2        chr1  BestRefSeq  transcript    685716    686654     .      -     .   \n",
       "3        chr1  BestRefSeq  transcript    923923    944574     .      +     .   \n",
       "4        chr1  BestRefSeq  transcript    944203    959256     .      -     .   \n",
       "...       ...         ...         ...       ...       ...   ...    ...   ...   \n",
       "19214    chrY  BestRefSeq  transcript  24618004  24639207     .      +     .   \n",
       "19215    chrY  BestRefSeq  transcript  24763069  24813393     .      -     .   \n",
       "19216    chrY  BestRefSeq  transcript  24833919  24907040     .      +     .   \n",
       "19217    chrY  BestRefSeq  transcript  25030901  25052104     .      -     .   \n",
       "19218    chrY  BestRefSeq  transcript  25622117  25624902     .      +     .   \n",
       "\n",
       "                                              attributes gene_id  \\\n",
       "0      gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...   OR4F5   \n",
       "1      gene_id \"OR4F29\"; transcript_id \"NM_001005221....  OR4F29   \n",
       "2      gene_id \"OR4F16\"; transcript_id \"NM_001005277....  OR4F16   \n",
       "3      gene_id \"SAMD11\"; transcript_id \"NM_001385641....  SAMD11   \n",
       "4      gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...   NOC2L   \n",
       "...                                                  ...     ...   \n",
       "19214  gene_id \"BPY2B\"; transcript_id \"NM_001002760.1...   BPY2B   \n",
       "19215  gene_id \"DAZ3\"; transcript_id \"NM_020364.4\"; d...    DAZ3   \n",
       "19216  gene_id \"DAZ4\"; transcript_id \"NM_001388484.1\"...    DAZ4   \n",
       "19217  gene_id \"BPY2C\"; transcript_id \"NM_001002761.1...   BPY2C   \n",
       "19218  gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...    CDY1   \n",
       "\n",
       "        transcript_id entrez_id  \n",
       "0      NM_001005484.2     79501  \n",
       "1      NM_001005221.2    729759  \n",
       "2      NM_001005277.1     81399  \n",
       "3      NM_001385641.1    148398  \n",
       "4         NM_015658.4     26155  \n",
       "...               ...       ...  \n",
       "19214  NM_001002760.1    442867  \n",
       "19215     NM_020364.4     57054  \n",
       "19216  NM_001388484.1     57135  \n",
       "19217  NM_001002761.1    442868  \n",
       "19218     NM_004680.3      9085  \n",
       "\n",
       "[19219 rows x 12 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_path = '/mnt/home/aaggarwal/ceph/gates_proj/ncbi_genome_hg38.p14/GCF_000001405.40_GRCh38.p14_genomic.final_transcript.gtf'\n",
    "\n",
    "col_names = [\n",
    "    \"seqname\", \"source\", \"feature\", \"start\", \"end\",\n",
    "    \"score\", \"strand\", \"frame\", \"attributes\"\n",
    "]\n",
    "\n",
    "gtf = pd.read_csv(\n",
    "    gtf_path,\n",
    "    sep='\\t',\n",
    "    comment='#',\n",
    "    names=col_names,\n",
    "    dtype=str\n",
    ")\n",
    "\n",
    "gtf['gene_id'] = gtf['attributes'].str.extract(r'gene_id \"([^\"]+)\"')\n",
    "gtf['transcript_id'] = gtf['attributes'].str.extract(r'transcript_id \"([^\"]+)\"')\n",
    "gtf['entrez_id'] = gtf['attributes'].str.extract(r'GeneID:(\\d+)')\n",
    "\n",
    "gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "c6a6520c-1dcf-480f-a713-e29b6dc7ebe0",
   "metadata": {},
   "outputs": [],
   "source": [
    "transcript_counts = gtf.groupby(\"gene_id\")[\"transcript_id\"].nunique()\n",
    "\n",
    "multiple_transcripts = transcript_counts[transcript_counts > 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "65fd83ed-e6c4-40cb-b2cc-beae88be3526",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: transcript_id, dtype: int64)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multiple_transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "4b01ed88-d9c3-4205-9b00-6d96888cdea2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19219"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_gene_ids_count = gtf[\"gene_id\"].nunique()\n",
    "unique_gene_ids_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "21f133b5-31fd-4c0b-89cd-a93ed54c317a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19219"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_gene_ids_count = gtf[\"entrez_id\"].nunique()\n",
    "unique_gene_ids_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "bdfdbb7f-87fa-416e-929b-6ceae2978dd0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19219"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_gene_ids_count = gtf[\"transcript_id\"].nunique()\n",
    "unique_gene_ids_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "a85652e0-d0d9-4fd5-b5c9-799dcc20a819",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of entries with MANE Select: 19219\n"
     ]
    }
   ],
   "source": [
    "mane_select_count = gtf['attributes'].str.contains('MANE Select', na=False).sum()\n",
    "print(\"Number of entries with MANE Select:\", mane_select_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb2d936f-6471-4ddd-bf16-de0539d2dccb",
   "metadata": {},
   "source": [
    "## Check the csv GTF files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "60642a5d-23d4-47af-9843-e7c3adf4c458",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_cds = pd.read_csv('/mnt/home/aaggarwal/ceph/gates_proj/ncbi_genome_hg38.p14/hg38.p14.ncbiRefSeq.CDS_final_gtf.csv')\n",
    "gtf_trans = pd.read_csv('/mnt/home/aaggarwal/ceph/gates_proj/ncbi_genome_hg38.p14/hg38.p14.ncbiRefSeq.transcript_final_gtf.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "054a09b0-45ed-4013-aa7f-d85949f6a7f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>65565</td>\n",
       "      <td>65573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>79501</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>69037</td>\n",
       "      <td>70005</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>79501</td>\n",
       "      <td>969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>65565</td>\n",
       "      <td>65567</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>79501</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>70006</td>\n",
       "      <td>70008</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>79501</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>450743</td>\n",
       "      <td>451678</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"OR4F29\"; transcript_id \"NM_001005221....</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>NM_001005221.2</td>\n",
       "      <td>729759</td>\n",
       "      <td>936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229586</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>25038098</td>\n",
       "      <td>25038100</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"BPY2C\"; transcript_id \"NM_001002761.1...</td>\n",
       "      <td>BPY2C</td>\n",
       "      <td>NM_001002761.1</td>\n",
       "      <td>442868</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229587</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>25622443</td>\n",
       "      <td>25624034</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>NM_004680.3</td>\n",
       "      <td>9085</td>\n",
       "      <td>1592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229588</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>25624455</td>\n",
       "      <td>25624524</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>NM_004680.3</td>\n",
       "      <td>9085</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229589</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>25622443</td>\n",
       "      <td>25622445</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>NM_004680.3</td>\n",
       "      <td>9085</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229590</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>25624525</td>\n",
       "      <td>25624527</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>NM_004680.3</td>\n",
       "      <td>9085</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>229591 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname      source      feature     start       end score strand  \\\n",
       "0         chr1  BestRefSeq          CDS     65565     65573     .      +   \n",
       "1         chr1  BestRefSeq          CDS     69037     70005     .      +   \n",
       "2         chr1  BestRefSeq  start_codon     65565     65567     .      +   \n",
       "3         chr1  BestRefSeq   stop_codon     70006     70008     .      +   \n",
       "4         chr1  BestRefSeq          CDS    450743    451678     .      -   \n",
       "...        ...         ...          ...       ...       ...   ...    ...   \n",
       "229586    chrY  BestRefSeq   stop_codon  25038098  25038100     .      -   \n",
       "229587    chrY  BestRefSeq          CDS  25622443  25624034     .      +   \n",
       "229588    chrY  BestRefSeq          CDS  25624455  25624524     .      +   \n",
       "229589    chrY  BestRefSeq  start_codon  25622443  25622445     .      +   \n",
       "229590    chrY  BestRefSeq   stop_codon  25624525  25624527     .      +   \n",
       "\n",
       "        frame                                         attributes gene_id  \\\n",
       "0           0  gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...   OR4F5   \n",
       "1           0  gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...   OR4F5   \n",
       "2           0  gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...   OR4F5   \n",
       "3           0  gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...   OR4F5   \n",
       "4           0  gene_id \"OR4F29\"; transcript_id \"NM_001005221....  OR4F29   \n",
       "...       ...                                                ...     ...   \n",
       "229586      0  gene_id \"BPY2C\"; transcript_id \"NM_001002761.1...   BPY2C   \n",
       "229587      0  gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...    CDY1   \n",
       "229588      1  gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...    CDY1   \n",
       "229589      0  gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...    CDY1   \n",
       "229590      0  gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...    CDY1   \n",
       "\n",
       "         transcript_id  entrez_id  length  \n",
       "0       NM_001005484.2      79501       9  \n",
       "1       NM_001005484.2      79501     969  \n",
       "2       NM_001005484.2      79501       3  \n",
       "3       NM_001005484.2      79501       3  \n",
       "4       NM_001005221.2     729759     936  \n",
       "...                ...        ...     ...  \n",
       "229586  NM_001002761.1     442868       3  \n",
       "229587     NM_004680.3       9085    1592  \n",
       "229588     NM_004680.3       9085      70  \n",
       "229589     NM_004680.3       9085       3  \n",
       "229590     NM_004680.3       9085       3  \n",
       "\n",
       "[229591 rows x 13 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_cds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "b9b43131-49a8-47d9-b872-f2d532b60c2d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>entrez_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>65419</td>\n",
       "      <td>71585</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>79501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>450740</td>\n",
       "      <td>451678</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"OR4F29\"; transcript_id \"NM_001005221....</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>NM_001005221.2</td>\n",
       "      <td>729759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>685716</td>\n",
       "      <td>686654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"OR4F16\"; transcript_id \"NM_001005277....</td>\n",
       "      <td>OR4F16</td>\n",
       "      <td>NM_001005277.1</td>\n",
       "      <td>81399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>923923</td>\n",
       "      <td>944574</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"SAMD11\"; transcript_id \"NM_001385641....</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>NM_001385641.1</td>\n",
       "      <td>148398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>944203</td>\n",
       "      <td>959256</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658.4</td>\n",
       "      <td>26155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19161</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>24618004</td>\n",
       "      <td>24639207</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"BPY2B\"; transcript_id \"NM_001002760.1...</td>\n",
       "      <td>BPY2B</td>\n",
       "      <td>NM_001002760.1</td>\n",
       "      <td>442867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19162</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>24763069</td>\n",
       "      <td>24813393</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"DAZ3\"; transcript_id \"NM_020364.4\"; d...</td>\n",
       "      <td>DAZ3</td>\n",
       "      <td>NM_020364.4</td>\n",
       "      <td>57054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19163</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>24833919</td>\n",
       "      <td>24907040</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"DAZ4\"; transcript_id \"NM_001388484.1\"...</td>\n",
       "      <td>DAZ4</td>\n",
       "      <td>NM_001388484.1</td>\n",
       "      <td>57135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19164</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>25030901</td>\n",
       "      <td>25052104</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"BPY2C\"; transcript_id \"NM_001002761.1...</td>\n",
       "      <td>BPY2C</td>\n",
       "      <td>NM_001002761.1</td>\n",
       "      <td>442868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19165</th>\n",
       "      <td>chrY</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>25622117</td>\n",
       "      <td>25624902</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>NM_004680.3</td>\n",
       "      <td>9085</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19166 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqname      source     feature     start       end score strand frame  \\\n",
       "0        chr1  BestRefSeq  transcript     65419     71585     .      +     .   \n",
       "1        chr1  BestRefSeq  transcript    450740    451678     .      -     .   \n",
       "2        chr1  BestRefSeq  transcript    685716    686654     .      -     .   \n",
       "3        chr1  BestRefSeq  transcript    923923    944574     .      +     .   \n",
       "4        chr1  BestRefSeq  transcript    944203    959256     .      -     .   \n",
       "...       ...         ...         ...       ...       ...   ...    ...   ...   \n",
       "19161    chrY  BestRefSeq  transcript  24618004  24639207     .      +     .   \n",
       "19162    chrY  BestRefSeq  transcript  24763069  24813393     .      -     .   \n",
       "19163    chrY  BestRefSeq  transcript  24833919  24907040     .      +     .   \n",
       "19164    chrY  BestRefSeq  transcript  25030901  25052104     .      -     .   \n",
       "19165    chrY  BestRefSeq  transcript  25622117  25624902     .      +     .   \n",
       "\n",
       "                                              attributes gene_id  \\\n",
       "0      gene_id \"OR4F5\"; transcript_id \"NM_001005484.2...   OR4F5   \n",
       "1      gene_id \"OR4F29\"; transcript_id \"NM_001005221....  OR4F29   \n",
       "2      gene_id \"OR4F16\"; transcript_id \"NM_001005277....  OR4F16   \n",
       "3      gene_id \"SAMD11\"; transcript_id \"NM_001385641....  SAMD11   \n",
       "4      gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...   NOC2L   \n",
       "...                                                  ...     ...   \n",
       "19161  gene_id \"BPY2B\"; transcript_id \"NM_001002760.1...   BPY2B   \n",
       "19162  gene_id \"DAZ3\"; transcript_id \"NM_020364.4\"; d...    DAZ3   \n",
       "19163  gene_id \"DAZ4\"; transcript_id \"NM_001388484.1\"...    DAZ4   \n",
       "19164  gene_id \"BPY2C\"; transcript_id \"NM_001002761.1...   BPY2C   \n",
       "19165  gene_id \"CDY1\"; transcript_id \"NM_004680.3\"; d...    CDY1   \n",
       "\n",
       "        transcript_id  entrez_id  \n",
       "0      NM_001005484.2      79501  \n",
       "1      NM_001005221.2     729759  \n",
       "2      NM_001005277.1      81399  \n",
       "3      NM_001385641.1     148398  \n",
       "4         NM_015658.4      26155  \n",
       "...               ...        ...  \n",
       "19161  NM_001002760.1     442867  \n",
       "19162     NM_020364.4      57054  \n",
       "19163  NM_001388484.1      57135  \n",
       "19164  NM_001002761.1     442868  \n",
       "19165     NM_004680.3       9085  \n",
       "\n",
       "[19166 rows x 12 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_trans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "c656702f-d964-4557-840c-fd2ed3d01f63",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19166"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_gene_ids_count = gtf_cds[\"gene_id\"].nunique()\n",
    "unique_gene_ids_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "b731efea-13fd-4b25-aa43-6af02534212d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19166"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_gene_ids_count = gtf_cds[\"entrez_id\"].nunique()\n",
    "unique_gene_ids_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "5f4f1cd9-d8a3-402d-8f46-eb73bc3345f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm\n",
    "tqdm.pandas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "4b749da2-37fc-4835-8ff1-a502f60d95a6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 19166/19166 [03:04<00:00, 103.71it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "cds_overlap_count\n",
       "1    19166\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter only start_codon and CDS\n",
    "start_codons = gtf_cds[gtf_cds['feature'] == 'start_codon'].copy()\n",
    "cds_regions = gtf_cds[gtf_cds['feature'] == 'CDS'].copy()\n",
    "\n",
    "# Ensure all coordinates are integers\n",
    "start_codons[\"start\"] = start_codons[\"start\"].astype(int)\n",
    "start_codons[\"end\"] = start_codons[\"end\"].astype(int)\n",
    "cds_regions[\"start\"] = cds_regions[\"start\"].astype(int)\n",
    "cds_regions[\"end\"] = cds_regions[\"end\"].astype(int)\n",
    "\n",
    "# For each start codon, count how many CDS regions it overlaps with\n",
    "def count_overlaps(row, cds):\n",
    "    subset = cds[cds['transcript_id'] == row['transcript_id']]\n",
    "    return ((subset['start'] <= row['end']) & (subset['end'] >= row['start'])).sum()\n",
    "\n",
    "start_codons['cds_overlap_count'] = start_codons.progress_apply(lambda row: count_overlaps(row, cds_regions), axis=1)\n",
    "\n",
    "# Count how many start codons have exactly 1 overlapping CDS\n",
    "overlap_counts = start_codons['cds_overlap_count'].value_counts().sort_index()\n",
    "overlap_counts\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "ebd63222-ef5e-4a66-a918-5992fe487493",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 19166/19166 [03:12<00:00, 99.72it/s] "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✔️ Valid stop codons: 99.92%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# Filter stop codons\n",
    "stop_codons = gtf_cds[gtf_cds['feature'] == 'stop_codon'].copy()\n",
    "\n",
    "# Ensure coordinates are ints\n",
    "stop_codons[\"start\"] = stop_codons[\"start\"].astype(int)\n",
    "stop_codons[\"end\"] = stop_codons[\"end\"].astype(int)\n",
    "\n",
    "# Preprocess CDS\n",
    "cds_regions = gtf_cds[gtf_cds['feature'] == 'CDS'].copy()\n",
    "cds_regions[\"start\"] = cds_regions[\"start\"].astype(int)\n",
    "cds_regions[\"end\"] = cds_regions[\"end\"].astype(int)\n",
    "\n",
    "# For each stop codon, check for:\n",
    "# - overlap with any CDS\n",
    "# - or being exactly adjacent to the last CDS (strand-aware)\n",
    "def is_valid_stop(row, cds):\n",
    "    tx = row['transcript_id']\n",
    "    strand = row['strand']\n",
    "    tx_cds = cds[cds['transcript_id'] == tx]\n",
    "    \n",
    "    # Check overlap\n",
    "    overlap = ((tx_cds['start'] <= row['end']) & (tx_cds['end'] >= row['start'])).any()\n",
    "    if overlap:\n",
    "        return True\n",
    "\n",
    "    # Strand-aware adjacency\n",
    "    if strand == '+':\n",
    "        last_cds_end = tx_cds['end'].max()\n",
    "        return row['start'] == last_cds_end + 1\n",
    "    elif strand == '-':\n",
    "        first_cds_start = tx_cds['start'].min()\n",
    "        return row['end'] == first_cds_start - 1\n",
    "\n",
    "    return False  # fallback\n",
    "\n",
    "from tqdm import tqdm\n",
    "tqdm.pandas()\n",
    "\n",
    "stop_codons['is_valid_stop'] = stop_codons.progress_apply(lambda row: is_valid_stop(row, cds_regions), axis=1)\n",
    "\n",
    "# Summary\n",
    "valid_pct = stop_codons['is_valid_stop'].mean() * 100\n",
    "print(f\"✔️ Valid stop codons: {valid_pct:.2f}%\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "9937c641-a419-4644-a07f-f210180b49a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Checking transcripts: 100%|██████████| 19166/19166 [03:26<00:00, 92.63it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Found 0 transcripts where start codon overlaps CDS with frame > 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "cds_df = gtf_cds[gtf_cds['feature'] == 'CDS']\n",
    "start_codon_df = gtf_cds[gtf_cds['feature'] == 'start_codon']\n",
    "\n",
    "# Collect suspicious transcripts\n",
    "mismatched_transcripts = []\n",
    "\n",
    "# Iterate through all unique transcripts with start codons\n",
    "for transcript_id in tqdm(start_codon_df['transcript_id'].unique(), desc=\"Checking transcripts\"):\n",
    "    start_rows = start_codon_df[start_codon_df['transcript_id'] == transcript_id]\n",
    "    cds_rows = cds_df[cds_df['transcript_id'] == transcript_id]\n",
    "\n",
    "    if start_rows.empty or cds_rows.empty:\n",
    "        continue\n",
    "\n",
    "    # Assume only one start_codon row per transcript (standard for well-formed GTFs)\n",
    "    start_row = start_rows.iloc[0]\n",
    "    start_coord = start_row['start']\n",
    "    strand = start_row['strand']\n",
    "\n",
    "    # Find overlapping CDS\n",
    "    overlapping = cds_rows[\n",
    "        (cds_rows['start'] <= start_coord) &\n",
    "        (cds_rows['end'] >= start_coord)\n",
    "    ]\n",
    "\n",
    "    if not overlapping.empty:\n",
    "        frame = overlapping.iloc[0]['frame']\n",
    "        if str(frame).isdigit() and int(frame) != 0:\n",
    "            mismatched_transcripts.append({\n",
    "                'transcript_id': transcript_id,\n",
    "                'frame': frame,\n",
    "                'cds_start': overlapping.iloc[0]['start'],\n",
    "                'cds_end': overlapping.iloc[0]['end'],\n",
    "                'start_codon_pos': start_coord,\n",
    "                'strand': strand\n",
    "            })\n",
    "\n",
    "# Show results\n",
    "print(f\"\\nFound {len(mismatched_transcripts)} transcripts where start codon overlaps CDS with frame > 0\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "4cd9f499-0142-40b0-ac45-2e5a31e8774f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mismatched_transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "id": "c649615d-802a-4e56-80b1-eeec52255d35",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_adjusted_cds_windows(transcript_id, chrom, strand, gtf_df):\n",
    "    # Filter relevant entries\n",
    "    transcript_df = gtf_df[gtf_df['transcript_id'] == transcript_id]\n",
    "    cds_df = transcript_df[transcript_df['feature'] == 'CDS']\n",
    "    start_df = transcript_df[transcript_df['feature'] == 'start_codon']\n",
    "    stop_df = transcript_df[transcript_df['feature'] == 'stop_codon']\n",
    "\n",
    "    if cds_df.empty or start_df.empty or stop_df.empty:\n",
    "        return None  # Missing critical annotations\n",
    "\n",
    "    # Get coordinates\n",
    "    start_codon_start = start_df.iloc[0]['start']\n",
    "    stop_codon_end = stop_df.iloc[0]['end']\n",
    "\n",
    "    # Keep only CDSs that intersect coding region\n",
    "    if strand == '+':\n",
    "        cds_df = cds_df[\n",
    "            (cds_df['end'] >= start_codon_start) &\n",
    "            (cds_df['start'] <= stop_codon_end)\n",
    "        ].sort_values(by='start')\n",
    "    else:\n",
    "        start_codon_end = start_df.iloc[0]['end']\n",
    "        stop_codon_start = stop_df.iloc[0]['start']\n",
    "    \n",
    "        cds_df = cds_df[\n",
    "            (cds_df['end'] >= stop_codon_start) &\n",
    "            (cds_df['start'] <= start_codon_end)\n",
    "        ].sort_values(by='end', ascending=False)\n",
    "\n",
    "    if cds_df.empty:\n",
    "        return None\n",
    "\n",
    "    # Build trimmed CDS windows\n",
    "    cds_windows = []\n",
    "    for i, row in cds_df.iterrows():\n",
    "        start = row['start']\n",
    "        end = row['end']\n",
    "\n",
    "        # First CDS: adjust start if needed\n",
    "        if strand == '+':\n",
    "            if start <= start_codon_start <= end:\n",
    "                start = start_codon_start\n",
    "        else:\n",
    "            if start <= start_codon_end <= end:\n",
    "                end = start_codon_end\n",
    "\n",
    "        # Last CDS: adjust end if needed\n",
    "        if strand == '+':\n",
    "            if start <= stop_codon_end <= end:\n",
    "                end = stop_codon_end\n",
    "        else:\n",
    "            if start <= stop_codon_start <= end:\n",
    "                start = stop_codon_start\n",
    "\n",
    "        # Sanity check\n",
    "        if end > start:\n",
    "            cds_windows.append((start, end, int(row['frame'])))\n",
    "\n",
    "    return cds_windows if cds_windows else None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "id": "ab5bdfe4-96d0-4ea1-b550-8cccbacd7886",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>539</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1919182</td>\n",
       "      <td>1919265</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...</td>\n",
       "      <td>TMEM52</td>\n",
       "      <td>NM_178545.4</td>\n",
       "      <td>339456</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>540</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1919045</td>\n",
       "      <td>1919088</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...</td>\n",
       "      <td>TMEM52</td>\n",
       "      <td>NM_178545.4</td>\n",
       "      <td>339456</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1918893</td>\n",
       "      <td>1918934</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...</td>\n",
       "      <td>TMEM52</td>\n",
       "      <td>NM_178545.4</td>\n",
       "      <td>339456</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>542</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1918253</td>\n",
       "      <td>1918431</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...</td>\n",
       "      <td>TMEM52</td>\n",
       "      <td>NM_178545.4</td>\n",
       "      <td>339456</td>\n",
       "      <td>179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1917885</td>\n",
       "      <td>1918162</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...</td>\n",
       "      <td>TMEM52</td>\n",
       "      <td>NM_178545.4</td>\n",
       "      <td>339456</td>\n",
       "      <td>278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>544</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1919263</td>\n",
       "      <td>1919265</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...</td>\n",
       "      <td>TMEM52</td>\n",
       "      <td>NM_178545.4</td>\n",
       "      <td>339456</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>545</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>1917882</td>\n",
       "      <td>1917884</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...</td>\n",
       "      <td>TMEM52</td>\n",
       "      <td>NM_178545.4</td>\n",
       "      <td>339456</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqname      source      feature    start      end score strand  frame  \\\n",
       "539    chr1  BestRefSeq          CDS  1919182  1919265     .      -      0   \n",
       "540    chr1  BestRefSeq          CDS  1919045  1919088     .      -      0   \n",
       "541    chr1  BestRefSeq          CDS  1918893  1918934     .      -      1   \n",
       "542    chr1  BestRefSeq          CDS  1918253  1918431     .      -      1   \n",
       "543    chr1  BestRefSeq          CDS  1917885  1918162     .      -      2   \n",
       "544    chr1  BestRefSeq  start_codon  1919263  1919265     .      -      0   \n",
       "545    chr1  BestRefSeq   stop_codon  1917882  1917884     .      -      0   \n",
       "\n",
       "                                            attributes gene_id transcript_id  \\\n",
       "539  gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...  TMEM52   NM_178545.4   \n",
       "540  gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...  TMEM52   NM_178545.4   \n",
       "541  gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...  TMEM52   NM_178545.4   \n",
       "542  gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...  TMEM52   NM_178545.4   \n",
       "543  gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...  TMEM52   NM_178545.4   \n",
       "544  gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...  TMEM52   NM_178545.4   \n",
       "545  gene_id \"TMEM52\"; transcript_id \"NM_178545.4\";...  TMEM52   NM_178545.4   \n",
       "\n",
       "     entrez_id  length  \n",
       "539     339456      84  \n",
       "540     339456      44  \n",
       "541     339456      42  \n",
       "542     339456     179  \n",
       "543     339456     278  \n",
       "544     339456       3  \n",
       "545     339456       3  "
      ]
     },
     "execution_count": 232,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_cds[gtf_cds['entrez_id'] == 339456]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "id": "fd0e5e55-0d7c-4248-ac01-3ceeae8bb3e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>415</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1574478</td>\n",
       "      <td>1574557</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...</td>\n",
       "      <td>SSU72</td>\n",
       "      <td>NM_014188.3</td>\n",
       "      <td>29101</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>416</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1564773</td>\n",
       "      <td>1564916</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...</td>\n",
       "      <td>SSU72</td>\n",
       "      <td>NM_014188.3</td>\n",
       "      <td>29101</td>\n",
       "      <td>144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>417</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1544863</td>\n",
       "      <td>1545002</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...</td>\n",
       "      <td>SSU72</td>\n",
       "      <td>NM_014188.3</td>\n",
       "      <td>29101</td>\n",
       "      <td>140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>418</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1543869</td>\n",
       "      <td>1543987</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...</td>\n",
       "      <td>SSU72</td>\n",
       "      <td>NM_014188.3</td>\n",
       "      <td>29101</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1542069</td>\n",
       "      <td>1542167</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...</td>\n",
       "      <td>SSU72</td>\n",
       "      <td>NM_014188.3</td>\n",
       "      <td>29101</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>420</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1574555</td>\n",
       "      <td>1574557</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...</td>\n",
       "      <td>SSU72</td>\n",
       "      <td>NM_014188.3</td>\n",
       "      <td>29101</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>1542066</td>\n",
       "      <td>1542068</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...</td>\n",
       "      <td>SSU72</td>\n",
       "      <td>NM_014188.3</td>\n",
       "      <td>29101</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqname      source      feature    start      end score strand  frame  \\\n",
       "415    chr1  BestRefSeq          CDS  1574478  1574557     .      -      0   \n",
       "416    chr1  BestRefSeq          CDS  1564773  1564916     .      -      1   \n",
       "417    chr1  BestRefSeq          CDS  1544863  1545002     .      -      1   \n",
       "418    chr1  BestRefSeq          CDS  1543869  1543987     .      -      2   \n",
       "419    chr1  BestRefSeq          CDS  1542069  1542167     .      -      0   \n",
       "420    chr1  BestRefSeq  start_codon  1574555  1574557     .      -      0   \n",
       "421    chr1  BestRefSeq   stop_codon  1542066  1542068     .      -      0   \n",
       "\n",
       "                                            attributes gene_id transcript_id  \\\n",
       "415  gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...   SSU72   NM_014188.3   \n",
       "416  gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...   SSU72   NM_014188.3   \n",
       "417  gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...   SSU72   NM_014188.3   \n",
       "418  gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...   SSU72   NM_014188.3   \n",
       "419  gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...   SSU72   NM_014188.3   \n",
       "420  gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...   SSU72   NM_014188.3   \n",
       "421  gene_id \"SSU72\"; transcript_id \"NM_014188.3\"; ...   SSU72   NM_014188.3   \n",
       "\n",
       "     entrez_id  length  \n",
       "415      29101      80  \n",
       "416      29101     144  \n",
       "417      29101     140  \n",
       "418      29101     119  \n",
       "419      29101      99  \n",
       "420      29101       3  \n",
       "421      29101       3  "
      ]
     },
     "execution_count": 216,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_cds[gtf_cds['transcript_id'] == 'NM_014188.3']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "id": "68ea8cc0-acf8-4d48-a205-5f76fb10f48d",
   "metadata": {},
   "outputs": [],
   "source": [
    "windows = get_adjusted_cds_windows(\"NM_014188.3\", \"chr1\", \"-\", gtf_cds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "id": "b7f90260-5489-42c3-a76c-0e2bed785b84",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(1574478, 1574557, 0),\n",
       " (1564773, 1564916, 1),\n",
       " (1544863, 1545002, 1),\n",
       " (1543869, 1543987, 2),\n",
       " (1542069, 1542167, 0)]"
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "windows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "aad4dfaa-eb9d-4b0c-9660-9f1c004e3cd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from pyfaidx import Fasta\n",
    "from Bio.Seq import Seq\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "id": "e9a4772a-8cbf-4fb3-b7f6-0b79cdf1c080",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_fasta_dir = \"/mnt/home/aaggarwal/ceph/gates_proj/ncbi_genome_hg38.p14\"\n",
    "\n",
    "# load the genome\n",
    "fasta_files = {\n",
    "    f.replace(\".fa\", \"\"): os.path.join(genome_fasta_dir, f)\n",
    "    for f in os.listdir(genome_fasta_dir) if f.endswith(\".fa\")\n",
    "}\n",
    "fasta_index = {k: Fasta(v) for k, v in fasta_files.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "id": "82d1e195-1285-4573-a2e7-c5457b278bfd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(1574478, 1574557, 0),\n",
       " (1564773, 1564916, 1),\n",
       " (1544863, 1545002, 1),\n",
       " (1543869, 1543987, 2),\n",
       " (1542069, 1542167, 0)]"
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds_windows = sorted(windows, key=lambda x: x[0], reverse=True)\n",
    "cds_windows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "id": "5c16ac2f-3ab8-426d-a4c9-d049b5598c2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = fasta_index['chr1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "id": "9ff18a50-3f50-4ef6-a428-bba536a36e54",
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom = genome['chr1'] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "id": "715138e5-7c0f-4a93-bd27-a8942d88fa87",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_parts = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "id": "8ce9fbf2-7ff8-4622-8f42-61cf28537ab6",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, (start, end, frame) in enumerate(cds_windows):\n",
    "    start = int(start)\n",
    "    end = int(end)\n",
    "    frame = int(frame)\n",
    "\n",
    "    exon_seq = chrom[start-1:end].seq.upper()\n",
    "\n",
    "    if i == 0:\n",
    "        exon_seq = exon_seq[frame:]\n",
    "\n",
    "    seq_parts.append(exon_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "id": "8a1a61b0-7aa0-46fa-b140-6470f56aa67d",
   "metadata": {},
   "outputs": [],
   "source": [
    "full_seq = ''.join(seq_parts[::-1])  # reverse exon order\n",
    "full_seq_rc = str(Seq(full_seq).reverse_complement())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "id": "27454f8a-8551-4779-96c7-56fe6d5c8767",
   "metadata": {},
   "outputs": [],
   "source": [
    "protein = str(Seq(full_seq_rc).translate(to_stop=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "id": "94fd8bb5-d358-422e-9023-cccb0b177cd7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First codon: ATG\n",
      "Protein start: MPSSPLRVAV\n",
      "Full protein: MPSSPLRVAVVCSSNQNRSMEAHNILSKRGFSVRSFGTGTHVKLPGPAPDKPNVYDFKTTYDQMYNDLLRKDKELYTQNGILHMLDRNKRIKPRPERFQNCKDLFDLILTCEERVYDQVVEDLNSREQETCQPVHVVNVDIQDNHEEATLGAFLICELCQCIQHTEDMENEIDELLQEFEEKSGRTFLHTVCFY\n"
     ]
    }
   ],
   "source": [
    "print(f\"First codon: {full_seq_rc[:3]}\")\n",
    "print(f\"Protein start: {protein[:10]}\")\n",
    "print(f\"Full protein: {protein}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "id": "6e1e7911-2472-4241-93c3-a1554b8f1f0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_parts = []\n",
    "for i, (start, end, frame) in enumerate(cds_windows):\n",
    "    start = int(start)\n",
    "    end = int(end)\n",
    "    frame = int(frame)\n",
    "\n",
    "    exon_seq = chrom[start-1:end].seq.upper()\n",
    "\n",
    "    exon_seq = exon_seq[frame:]\n",
    "\n",
    "    seq_parts.append(exon_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "id": "3017411f-9c13-447e-ad9e-6e86940f3331",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First codon: ATG\n",
      "Protein start: MPSSPLRVAV\n",
      "Full protein: MPSSPLRVAVVCSSNQNRSMEAHNILSKRGFSVRSFGTGTHVKLPGPAPDKPNVYDFKTTYDQMYNDLLRKDKEPIHRMGFYICWTEIRESSPGQKDSRTAKTCLI*SSLAKRECMTRWWKSEFQRTGDLPACARGQCGHPGQPRGGHPGGVSHL*ALPVSSTRKTWRTRSTSCCRSSRRRVAAPFCTPSAS\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/home/aaggarwal/miniconda3/envs/target-genes/lib/python3.10/site-packages/Bio/Seq.py:2879: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "full_seq = ''.join(seq_parts[::-1])  # reverse exon order\n",
    "full_seq_rc = str(Seq(full_seq).reverse_complement())\n",
    "\n",
    "protein = str(Seq(full_seq_rc).translate(to_stop=False))\n",
    "\n",
    "print(f\"First codon: {full_seq_rc[:3]}\")\n",
    "print(f\"Protein start: {protein[:10]}\")\n",
    "print(f\"Full protein: {protein}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "791eb40d-a6a0-4bd5-aee0-3ce998154f93",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = fasta_index['chr1']\n",
    "\n",
    "chrom = genome['chr1']  # OR just use fasta_index['chr1']\n",
    "\n",
    "# Now slice the sequence: note that pyfaidx uses 0-based, end-inclusive indexing\n",
    "seq = chrom[1407215 - 1 : 1407217].seq.upper()\n",
    "rc = str(Seq(seq).reverse_complement())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "123d0e0c-376f-4d5e-9ce0-71c4b78385f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ATG'"
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "8b063c9d-6e6c-47d1-81a0-a962712b7721",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1071820</td>\n",
       "      <td>1072566</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"RNF223\"; transcript_id \"NM_001205252....</td>\n",
       "      <td>RNF223</td>\n",
       "      <td>NM_001205252.2</td>\n",
       "      <td>401934</td>\n",
       "      <td>747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1072564</td>\n",
       "      <td>1072566</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"RNF223\"; transcript_id \"NM_001205252....</td>\n",
       "      <td>RNF223</td>\n",
       "      <td>NM_001205252.2</td>\n",
       "      <td>401934</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>1071817</td>\n",
       "      <td>1071819</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"RNF223\"; transcript_id \"NM_001205252....</td>\n",
       "      <td>RNF223</td>\n",
       "      <td>NM_001205252.2</td>\n",
       "      <td>401934</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqname      source      feature    start      end score strand  frame  \\\n",
       "132    chr1  BestRefSeq          CDS  1071820  1072566     .      -      0   \n",
       "133    chr1  BestRefSeq  start_codon  1072564  1072566     .      -      0   \n",
       "134    chr1  BestRefSeq   stop_codon  1071817  1071819     .      -      0   \n",
       "\n",
       "                                            attributes gene_id  \\\n",
       "132  gene_id \"RNF223\"; transcript_id \"NM_001205252....  RNF223   \n",
       "133  gene_id \"RNF223\"; transcript_id \"NM_001205252....  RNF223   \n",
       "134  gene_id \"RNF223\"; transcript_id \"NM_001205252....  RNF223   \n",
       "\n",
       "      transcript_id  entrez_id  length  \n",
       "132  NM_001205252.2     401934     747  \n",
       "133  NM_001205252.2     401934       3  \n",
       "134  NM_001205252.2     401934       3  "
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_cds[gtf_cds['transcript_id'] == 'NM_001205252.2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "914a02ce-1bb7-4b4b-b969-030bcc84c56c",
   "metadata": {},
   "outputs": [],
   "source": [
    "windows = get_adjusted_cds_windows(\"NM_001205252.2\", \"chr1\", \"-\", gtf_cds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "05661300-6dd0-43f6-b070-aebc264fd552",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(1071820, 1072566, 0)]"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "windows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "c1937a08-7e99-469c-8077-6b47098951ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Scanning transcripts: 100%|██████████| 19166/19166 [03:48<00:00, 83.95it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Transcript with start codon in middle of CDS: ['NM_001005221.2']\n",
      "Transcript with stop codon in middle of CDS: []\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "from tqdm import tqdm\n",
    "\n",
    "start_codons = gtf_cds[gtf_cds['feature'] == 'start_codon']\n",
    "stop_codons = gtf_cds[gtf_cds['feature'] == 'stop_codon']\n",
    "cds = gtf_cds[gtf_cds['feature'] == 'CDS']\n",
    "\n",
    "start_in_middle = []\n",
    "stop_in_middle = []\n",
    "\n",
    "for transcript_id in tqdm(gtf_cds['transcript_id'].unique(), desc=\"Scanning transcripts\"):\n",
    "    cds_rows = cds[cds['transcript_id'] == transcript_id].sort_values(by='start')\n",
    "    if cds_rows.empty:\n",
    "        continue\n",
    "\n",
    "    start_row = start_codons[start_codons['transcript_id'] == transcript_id]\n",
    "    stop_row = stop_codons[stop_codons['transcript_id'] == transcript_id]\n",
    "    if start_row.empty or stop_row.empty:\n",
    "        continue\n",
    "\n",
    "    start_pos = start_row.iloc[0]['start']\n",
    "    stop_pos = stop_row.iloc[0]['end']\n",
    "    strand = cds_rows.iloc[0]['strand']\n",
    "\n",
    "    if strand == '+':\n",
    "        if start_pos > cds_rows.iloc[0]['start']:\n",
    "            start_in_middle.append(transcript_id)\n",
    "        if stop_pos < cds_rows.iloc[-1]['end']:\n",
    "            stop_in_middle.append(transcript_id)\n",
    "    else:\n",
    "        if start_pos < cds_rows.iloc[-1]['end']:\n",
    "            start_in_middle.append(transcript_id)\n",
    "        if stop_pos > cds_rows.iloc[0]['start']:\n",
    "            stop_in_middle.append(transcript_id)\n",
    "\n",
    "print(\"Transcript with start codon in middle of CDS:\", start_in_middle[:1])\n",
    "print(\"Transcript with stop codon in middle of CDS:\", stop_in_middle[:1])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "e5e83eef-43cf-4494-b30a-fb7f93b5e3ae",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['NM_001005221.2',\n",
       " 'NM_001005277.1',\n",
       " 'NM_015658.4',\n",
       " 'NM_001394713.1',\n",
       " 'NM_021170.4',\n",
       " 'NM_001205252.2',\n",
       " 'NM_017891.5',\n",
       " 'NM_004195.3',\n",
       " 'NM_003327.4',\n",
       " 'NM_016176.6',\n",
       " 'NM_001014980.3',\n",
       " 'NM_058167.3',\n",
       " 'NM_030649.3',\n",
       " 'NM_017871.6',\n",
       " 'NM_001330311.2',\n",
       " 'NM_032348.4',\n",
       " 'NM_017900.3',\n",
       " 'NM_030937.6',\n",
       " 'NM_017971.4',\n",
       " 'NM_001145210.3',\n",
       " 'NM_001114748.2',\n",
       " 'NM_014188.3',\n",
       " 'NM_001242659.2',\n",
       " 'NM_033486.3',\n",
       " 'NM_001290264.2',\n",
       " 'NM_024011.4',\n",
       " 'NM_023018.5',\n",
       " 'NM_002074.5',\n",
       " 'NM_178545.4',\n",
       " 'NM_001304360.2',\n",
       " 'NM_182533.4',\n",
       " 'NM_024848.3',\n",
       " 'NM_002617.4',\n",
       " 'NM_018216.4',\n",
       " 'NM_001010926.4',\n",
       " 'NM_033467.4',\n",
       " 'NM_001242672.3',\n",
       " 'NM_001409.4',\n",
       " 'NM_017818.4',\n",
       " 'NM_020710.3',\n",
       " 'NM_014704.4',\n",
       " 'NM_207356.3',\n",
       " 'NM_015102.5',\n",
       " 'NM_015557.3',\n",
       " 'NM_000983.4',\n",
       " 'NM_012405.4',\n",
       " 'NM_207370.4',\n",
       " 'NM_007274.4',\n",
       " 'NM_019089.5',\n",
       " 'NM_003790.3',\n",
       " 'NM_020631.6',\n",
       " 'NM_024654.5',\n",
       " 'NM_014851.4',\n",
       " 'NM_018198.4',\n",
       " 'NM_006786.4',\n",
       " 'NM_001561.6',\n",
       " 'NM_018948.4',\n",
       " 'NM_001042681.2',\n",
       " 'NM_001428.5',\n",
       " 'NM_207420.3',\n",
       " 'NM_003039.3',\n",
       " 'NM_024980.5',\n",
       " 'NM_001009566.3',\n",
       " 'NM_020248.3',\n",
       " 'NM_032368.5',\n",
       " 'NM_004401.3',\n",
       " 'NM_001079843.3',\n",
       " 'NM_001170754.2',\n",
       " 'NM_006610.4',\n",
       " 'NM_003132.3',\n",
       " 'NM_001001998.3',\n",
       " 'NM_004958.4',\n",
       " 'NM_012168.6',\n",
       " 'NM_006341.4',\n",
       " 'NM_005957.5',\n",
       " 'NM_006172.4',\n",
       " 'NM_002521.3',\n",
       " 'NM_138346.3',\n",
       " 'NM_004753.7',\n",
       " 'NM_001146344.3',\n",
       " 'NM_001013631.3',\n",
       " 'NM_001009611.4',\n",
       " 'NM_001039361.4',\n",
       " 'NM_001010889.2',\n",
       " 'NM_001300891.2',\n",
       " 'NM_001136561.3',\n",
       " 'NM_001306072.3',\n",
       " 'NM_001302551.2',\n",
       " 'NM_001291380.1',\n",
       " 'NM_001099850.2',\n",
       " 'NM_001012276.3',\n",
       " 'NM_001024661.2',\n",
       " 'NM_001099790.5',\n",
       " 'NM_001010847.2',\n",
       " 'NM_001229.5',\n",
       " 'NM_024758.5',\n",
       " 'NM_001089591.2',\n",
       " 'NM_003443.3',\n",
       " 'NM_014424.5',\n",
       " 'NM_182623.3',\n",
       " 'NM_004431.5',\n",
       " 'NM_153213.5',\n",
       " 'NM_030907.4',\n",
       " 'NM_018994.3',\n",
       " 'NM_198546.1',\n",
       " 'NM_002403.4',\n",
       " 'NM_022089.4',\n",
       " 'NM_003000.3',\n",
       " 'NM_007365.3',\n",
       " 'NM_018715.4',\n",
       " 'NM_152232.6',\n",
       " 'NM_003748.4',\n",
       " 'NM_001136265.2',\n",
       " 'NM_020765.3',\n",
       " 'NM_015047.3',\n",
       " 'NM_012067.3',\n",
       " 'NM_003689.4',\n",
       " 'NM_004930.5',\n",
       " 'NM_181719.7',\n",
       " 'NM_019062.2',\n",
       " 'NM_014589.3',\n",
       " 'NM_001395463.1',\n",
       " 'NM_012400.4',\n",
       " 'NM_001367969.2',\n",
       " 'NM_018584.6',\n",
       " 'NM_024544.3',\n",
       " 'NM_005216.5',\n",
       " 'NM_001122819.3',\n",
       " 'NM_001103161.2',\n",
       " 'NM_001372052.1',\n",
       " 'NM_001391906.1',\n",
       " 'NM_001397.3',\n",
       " 'NM_002885.4',\n",
       " 'NM_032236.8',\n",
       " 'NM_005529.7',\n",
       " 'NM_030761.5',\n",
       " 'NM_001289974.2',\n",
       " 'NM_001395462.2',\n",
       " 'NM_000864.5',\n",
       " 'NM_005826.5',\n",
       " 'NM_001077195.2',\n",
       " 'NM_003196.3',\n",
       " 'NM_017707.4',\n",
       " 'NM_004091.4',\n",
       " 'NM_002167.5',\n",
       " 'NM_001008216.2',\n",
       " 'NM_000191.3',\n",
       " 'NM_000147.5',\n",
       " 'NM_001841.3',\n",
       " 'NM_054016.4',\n",
       " 'NM_152372.4',\n",
       " 'NM_021258.4',\n",
       " 'NM_170743.4',\n",
       " 'NM_001199013.2',\n",
       " 'NM_004350.3',\n",
       " 'NM_015484.5',\n",
       " 'NM_020317.5',\n",
       " 'NM_020485.8',\n",
       " 'NM_024037.3',\n",
       " 'NM_178422.6',\n",
       " 'NM_005563.4',\n",
       " 'NM_000437.4',\n",
       " 'NM_001004434.3',\n",
       " 'NM_032588.4',\n",
       " 'NM_001364669.2',\n",
       " 'NM_001395468.1',\n",
       " 'NM_001389556.1',\n",
       " 'NM_001039775.4',\n",
       " 'NM_001114759.3',\n",
       " 'NM_018066.4',\n",
       " 'NM_022078.3',\n",
       " 'NM_021969.3',\n",
       " 'NM_152365.3',\n",
       " 'NM_052943.4',\n",
       " 'NM_003047.5',\n",
       " 'NM_004672.5',\n",
       " 'NM_003665.4',\n",
       " 'NM_001330448.1',\n",
       " 'NM_006990.5',\n",
       " 'NM_001371928.1',\n",
       " 'NM_005248.3',\n",
       " 'NM_002038.4',\n",
       " 'NM_002946.5',\n",
       " 'NM_001990.4',\n",
       " 'NM_000952.5',\n",
       " 'NM_014280.3',\n",
       " 'NM_005644.4',\n",
       " 'NM_001003682.4',\n",
       " 'NM_005626.5',\n",
       " 'NM_016011.5',\n",
       " 'NM_002379.3',\n",
       " 'NM_006762.3',\n",
       " 'NM_014654.4',\n",
       " 'NM_001020658.2',\n",
       " 'NM_024522.3',\n",
       " 'NM_004814.3',\n",
       " 'NM_004102.5',\n",
       " 'NM_012392.4',\n",
       " 'NM_001856.4',\n",
       " 'NM_001364857.2',\n",
       " 'NM_144569.7',\n",
       " 'NM_080391.4',\n",
       " 'NM_001414727.1',\n",
       " 'NM_019118.5',\n",
       " 'NM_023009.7',\n",
       " 'NM_001167676.2',\n",
       " 'NM_018045.8',\n",
       " 'NM_178547.5',\n",
       " 'NM_030786.3',\n",
       " 'NM_003680.4',\n",
       " 'NM_153756.3',\n",
       " 'NM_033504.4',\n",
       " 'NM_001300826.2',\n",
       " 'NM_001625.4',\n",
       " 'NM_018207.3',\n",
       " 'NM_001080438.1',\n",
       " 'NM_001385109.1',\n",
       " 'NM_001281956.2',\n",
       " 'NM_138428.6',\n",
       " 'NM_001080418.3',\n",
       " 'NM_001195156.2',\n",
       " 'NM_007167.4',\n",
       " 'NM_005066.3',\n",
       " 'NM_024874.5',\n",
       " 'NM_002794.5',\n",
       " 'NM_152374.2',\n",
       " 'NM_022111.4',\n",
       " 'NM_005202.4',\n",
       " 'NM_014408.5',\n",
       " 'NM_001304762.2',\n",
       " 'NM_001282547.2',\n",
       " 'NM_032881.3',\n",
       " 'NM_145047.5',\n",
       " 'NM_031280.4',\n",
       " 'NM_000760.4',\n",
       " 'NM_000831.4',\n",
       " 'NM_001270875.3',\n",
       " 'NM_024700.4',\n",
       " 'NM_013285.3',\n",
       " 'NM_001242908.2',\n",
       " 'NM_001350770.2',\n",
       " 'NM_001099439.2',\n",
       " 'NM_024640.4',\n",
       " 'NM_005955.3',\n",
       " 'NM_005540.3',\n",
       " 'NM_006802.4',\n",
       " 'NM_004468.5',\n",
       " 'NM_002699.4',\n",
       " 'NM_022157.4',\n",
       " 'NM_012333.5',\n",
       " 'NM_030772.5',\n",
       " 'NM_017821.5',\n",
       " 'NM_001135653.2',\n",
       " 'NM_014571.4',\n",
       " 'NM_032526.3',\n",
       " 'NM_016257.4',\n",
       " 'NM_001720.5',\n",
       " 'NM_022120.2',\n",
       " 'NM_017646.6',\n",
       " 'NM_001033081.3',\n",
       " 'NM_000310.4',\n",
       " 'NM_001852.4',\n",
       " 'NM_014747.3',\n",
       " 'NM_133467.3',\n",
       " 'NM_144990.4',\n",
       " 'NM_001394311.1',\n",
       " 'NM_001956.5',\n",
       " 'NM_024503.5',\n",
       " 'NM_033553.3',\n",
       " 'NM_014947.5',\n",
       " 'NM_032257.5',\n",
       " 'NM_148960.3',\n",
       " 'NM_022356.4',\n",
       " 'NM_199342.4',\n",
       " 'NM_006516.4',\n",
       " 'NM_006824.3',\n",
       " 'NM_182517.3',\n",
       " 'NM_022821.4',\n",
       " 'NM_201542.5',\n",
       " 'NM_001190880.3',\n",
       " 'NM_001024845.3',\n",
       " 'NM_001358438.1',\n",
       " 'NM_024066.3',\n",
       " 'NM_024587.4',\n",
       " 'NM_153274.3',\n",
       " 'NM_001377534.1',\n",
       " 'NM_003738.5',\n",
       " 'NM_020365.5',\n",
       " 'NM_024602.6',\n",
       " 'NM_020883.2',\n",
       " 'NM_001048174.2',\n",
       " 'NM_007170.3',\n",
       " 'NM_001102601.3',\n",
       " 'NM_181697.3',\n",
       " 'NM_001114938.3',\n",
       " 'NM_021639.5',\n",
       " 'NM_005897.3',\n",
       " 'NM_003629.4',\n",
       " 'NM_001328655.2',\n",
       " 'NM_017739.4',\n",
       " 'NM_006369.5',\n",
       " 'NM_001396071.1',\n",
       " 'NM_001322255.2',\n",
       " 'NM_001135553.4',\n",
       " 'NM_201403.3',\n",
       " 'NM_001394565.1',\n",
       " 'NM_014774.3',\n",
       " 'NM_000778.4',\n",
       " 'NM_005764.4',\n",
       " 'NM_001290403.2',\n",
       " 'NM_001048166.1',\n",
       " 'NM_001194986.2',\n",
       " 'NM_019073.4',\n",
       " 'NM_032785.4',\n",
       " 'NM_024603.4',\n",
       " 'NM_032110.3',\n",
       " 'NM_007051.3',\n",
       " 'NM_001297663.2',\n",
       " 'NM_001981.3',\n",
       " 'NM_001101662.2',\n",
       " 'NM_002867.4',\n",
       " 'NM_015913.4',\n",
       " 'NM_138417.3',\n",
       " 'NM_001330585.2',\n",
       " 'NM_004153.4',\n",
       " 'NM_001009881.3',\n",
       " 'NM_023077.3',\n",
       " 'NM_001198961.2',\n",
       " 'NM_006671.6',\n",
       " 'NM_017887.3',\n",
       " 'NM_002370.4',\n",
       " 'NM_004631.5',\n",
       " 'NM_001367484.1',\n",
       " 'NM_018087.5',\n",
       " 'NM_018982.5',\n",
       " 'NM_016126.4',\n",
       " 'NM_001010978.4',\n",
       " 'NM_004872.5',\n",
       " 'NM_001353655.3',\n",
       " 'NM_001031672.4',\n",
       " 'NM_145716.4',\n",
       " 'NM_176782.3',\n",
       " 'NM_152268.4',\n",
       " 'NM_001114108.2',\n",
       " 'NM_014762.4',\n",
       " 'NM_015306.3',\n",
       " 'NM_003713.5',\n",
       " 'NM_001004303.5',\n",
       " 'NM_000066.4',\n",
       " 'NM_001365792.1',\n",
       " 'NM_145243.5',\n",
       " 'NM_002353.3',\n",
       " 'NM_001085487.3',\n",
       " 'NM_002228.4',\n",
       " 'NM_000775.4',\n",
       " 'NM_152377.3',\n",
       " 'NM_032027.3',\n",
       " 'NM_181712.5',\n",
       " 'NM_001367561.1',\n",
       " 'NM_014288.5',\n",
       " 'NM_002227.4',\n",
       " 'NM_005478.6',\n",
       " 'NM_024763.5',\n",
       " 'NM_015139.3',\n",
       " 'NM_001276351.2',\n",
       " 'NM_001018069.2',\n",
       " 'NM_018841.6',\n",
       " 'NM_004675.5',\n",
       " 'NM_024911.7',\n",
       " 'NM_000329.3',\n",
       " 'NM_001114120.3',\n",
       " 'NM_017768.5',\n",
       " 'NM_030816.5',\n",
       " 'NM_198719.2',\n",
       " 'NM_203350.3',\n",
       " 'NM_173808.3',\n",
       " 'NM_001105659.2',\n",
       " 'NM_001382280.1',\n",
       " 'NM_001002912.5',\n",
       " 'NM_001889.4',\n",
       " 'NM_001130058.2',\n",
       " 'NM_080868.3',\n",
       " 'NM_005482.3',\n",
       " 'NM_015534.6',\n",
       " 'NM_201624.3',\n",
       " 'NM_003902.5',\n",
       " 'NM_022159.4',\n",
       " 'NM_024686.6',\n",
       " 'NM_005274.3',\n",
       " 'NM_004388.3',\n",
       " 'NM_001166293.2',\n",
       " 'NM_012152.3',\n",
       " 'NM_153259.4',\n",
       " 'NM_018298.11',\n",
       " 'NM_032184.2',\n",
       " 'NM_198077.4',\n",
       " 'NM_003921.5',\n",
       " 'NM_012137.4',\n",
       " 'NM_017953.4',\n",
       " 'NM_152890.7',\n",
       " 'NM_001366781.1',\n",
       " 'NM_004261.5',\n",
       " 'NM_001514.6',\n",
       " 'NM_001008661.3',\n",
       " 'NM_001162536.3',\n",
       " 'NM_018284.3',\n",
       " 'NM_002053.3',\n",
       " 'NM_004120.5',\n",
       " 'NM_207398.3',\n",
       " 'NM_052941.5',\n",
       " 'NM_052942.5',\n",
       " 'NM_020063.2',\n",
       " 'NM_201269.3',\n",
       " 'NM_001017975.6',\n",
       " 'NM_003243.5',\n",
       " 'NM_001287737.2',\n",
       " 'NM_053274.3',\n",
       " 'NM_005263.5',\n",
       " 'NM_001350197.2',\n",
       " 'NM_001006605.5',\n",
       " 'NM_016040.5',\n",
       " 'NM_003567.4',\n",
       " 'NM_014597.5',\n",
       " 'NM_002061.4',\n",
       " 'NM_000350.3',\n",
       " 'NM_004815.4',\n",
       " 'NM_001993.5',\n",
       " 'NM_001839.5',\n",
       " 'NM_144988.4',\n",
       " 'NM_000110.4',\n",
       " 'NM_001037317.2',\n",
       " 'NM_001361041.2',\n",
       " 'NM_194292.3',\n",
       " 'NM_144620.4',\n",
       " 'NM_001918.5',\n",
       " 'NM_001033025.3',\n",
       " 'NM_015958.3',\n",
       " 'NM_058170.4',\n",
       " 'NM_001854.4',\n",
       " 'NM_001008218.2',\n",
       " 'NM_006113.5',\n",
       " 'NM_013386.5',\n",
       " 'NM_001143989.3',\n",
       " 'NM_001102592.2',\n",
       " 'NM_152763.5',\n",
       " 'NM_001377458.1',\n",
       " 'NM_001142551.2',\n",
       " 'NM_005645.4',\n",
       " 'NM_001245025.3',\n",
       " 'NM_001032291.3',\n",
       " 'NM_001010985.3',\n",
       " 'NM_002959.7',\n",
       " 'NM_002790.4',\n",
       " 'NM_020703.4',\n",
       " 'NM_001377295.2',\n",
       " 'NM_000849.5',\n",
       " 'NM_133181.4',\n",
       " 'NM_006492.3',\n",
       " 'NM_004696.3',\n",
       " 'NM_001382293.1',\n",
       " 'NM_005549.2',\n",
       " 'NM_004974.4',\n",
       " 'NM_002232.5',\n",
       " 'NM_018372.4',\n",
       " 'NM_001349884.2',\n",
       " 'NM_024901.5',\n",
       " 'NM_002557.4',\n",
       " 'NM_024102.4',\n",
       " 'NM_020683.7',\n",
       " 'NM_019099.5',\n",
       " 'NM_001378969.1',\n",
       " 'NM_017744.5',\n",
       " 'NM_175744.5',\n",
       " 'NM_005167.7',\n",
       " 'NM_003051.4',\n",
       " 'NM_001323043.2',\n",
       " 'NM_018364.5',\n",
       " 'NM_015967.8',\n",
       " 'NM_001010922.3',\n",
       " 'NM_001253852.3',\n",
       " 'NM_001253772.2',\n",
       " 'NM_015906.4',\n",
       " 'NM_005872.3',\n",
       " 'NM_001256404.2',\n",
       " 'NM_000036.3',\n",
       " 'NM_002524.5',\n",
       " 'NM_001007553.3',\n",
       " 'NM_025073.3',\n",
       " 'NM_005725.6',\n",
       " 'NM_002506.3',\n",
       " 'NM_001232.4',\n",
       " 'NM_005599.3',\n",
       " 'NM_001779.3',\n",
       " 'NM_001007237.3',\n",
       " 'NM_025188.4',\n",
       " 'NM_024626.4',\n",
       " 'NM_017686.4',\n",
       " 'NM_206996.4',\n",
       " 'NM_001330677.2',\n",
       " 'NM_015836.4',\n",
       " 'NM_001080470.2',\n",
       " 'NM_005518.4',\n",
       " 'NM_032044.4',\n",
       " 'NM_021794.4',\n",
       " 'NM_024408.4',\n",
       " 'NM_004892.6',\n",
       " 'NM_001100910.2',\n",
       " 'NM_001372105.2',\n",
       " 'NM_001287385.2',\n",
       " 'NM_001144032.3',\n",
       " 'NM_001385408.1',\n",
       " 'NM_001271870.5',\n",
       " 'NM_001164261.2',\n",
       " 'NM_001397211.1',\n",
       " 'NM_001201325.2',\n",
       " 'NM_014455.4',\n",
       " 'NM_006099.3',\n",
       " 'NM_144698.5',\n",
       " 'NM_003637.5',\n",
       " 'NM_003846.3',\n",
       " 'NM_005105.5',\n",
       " 'NM_153713.3',\n",
       " 'NM_001039888.4',\n",
       " 'NM_006472.6',\n",
       " 'NM_213653.4',\n",
       " 'NM_001302371.3',\n",
       " 'NM_203458.6',\n",
       " 'NM_001368128.3',\n",
       " 'NM_005399.5',\n",
       " 'NM_001461.4',\n",
       " 'NM_016361.5',\n",
       " 'NM_181703.4',\n",
       " 'NM_001385469.3',\n",
       " 'NM_001123068.3',\n",
       " 'NM_001395631.1',\n",
       " 'NM_001364008.2',\n",
       " 'NM_001388367.1',\n",
       " 'NM_001024599.5',\n",
       " 'NM_001123375.3',\n",
       " 'NM_021059.3',\n",
       " 'NM_003516.3',\n",
       " 'NM_001034077.4',\n",
       " 'NM_003528.3',\n",
       " 'NM_175065.3',\n",
       " 'NM_014849.5',\n",
       " 'NM_005850.5',\n",
       " 'NM_001145862.2',\n",
       " 'NM_020205.4',\n",
       " 'NM_030920.5',\n",
       " 'NM_001077628.3',\n",
       " 'NM_021960.5',\n",
       " 'NM_004436.4',\n",
       " 'NM_018178.6',\n",
       " 'NM_032132.5',\n",
       " 'NM_004079.5',\n",
       " 'NM_000396.4',\n",
       " 'NM_001668.4',\n",
       " 'NM_022075.5',\n",
       " 'NM_001376665.1',\n",
       " 'NM_020239.4',\n",
       " 'NM_030913.6',\n",
       " 'NM_212551.5',\n",
       " 'NM_013353.3',\n",
       " 'NM_005997.3',\n",
       " 'NM_001369623.2',\n",
       " 'NM_001025603.2',\n",
       " 'NM_003944.4',\n",
       " 'NM_015100.4',\n",
       " 'NM_007185.7',\n",
       " 'NM_031420.4',\n",
       " 'NM_001083965.2',\n",
       " 'NM_001004432.4',\n",
       " 'NM_005060.4',\n",
       " 'NM_001394591.1',\n",
       " 'NM_182578.4',\n",
       " 'NM_053055.5',\n",
       " 'NM_002966.3',\n",
       " 'NM_005620.2',\n",
       " 'NM_001008536.2',\n",
       " 'NM_007113.4',\n",
       " 'NM_001122965.1',\n",
       " 'NM_001009931.3',\n",
       " 'NM_002016.2',\n",
       " 'NM_001014342.3',\n",
       " 'NM_016190.3',\n",
       " 'NM_178435.4',\n",
       " 'NM_032563.2',\n",
       " 'NM_178431.1',\n",
       " 'NM_178351.4',\n",
       " 'NM_006945.5',\n",
       " 'NM_005988.3',\n",
       " 'NM_001388198.1',\n",
       " 'NM_001024209.4',\n",
       " 'NM_001014450.3',\n",
       " 'NM_001014291.4',\n",
       " 'NM_052891.3',\n",
       " 'NM_020393.4',\n",
       " 'NM_005621.2',\n",
       " 'NM_002964.5',\n",
       " 'NM_002963.4',\n",
       " 'NM_014624.4',\n",
       " 'NM_001394232.1',\n",
       " 'NM_002961.3',\n",
       " 'NM_002960.2',\n",
       " 'NM_005978.4',\n",
       " 'NM_080388.3',\n",
       " 'NM_020672.3',\n",
       " 'NM_001024211.2',\n",
       " 'NM_004515.4',\n",
       " 'NM_020699.4',\n",
       " 'NM_014856.3',\n",
       " 'NM_181715.3',\n",
       " 'NM_001271958.2',\n",
       " 'NM_006694.4',\n",
       " 'NM_002870.5',\n",
       " 'NM_207308.3',\n",
       " 'NM_152263.4',\n",
       " 'NM_001010979.3',\n",
       " 'NM_001098616.3',\n",
       " 'NM_001010846.3',\n",
       " 'NM_017582.7',\n",
       " 'NM_001111.5',\n",
       " 'NM_002249.6',\n",
       " 'NM_006556.4',\n",
       " 'NM_020524.4',\n",
       " 'NM_138300.4',\n",
       " 'NM_001130040.2',\n",
       " 'NM_144622.3',\n",
       " 'NM_153741.2',\n",
       " 'NM_173852.4',\n",
       " 'NM_007112.5',\n",
       " 'NM_000157.4',\n",
       " 'NM_006589.3',\n",
       " 'NM_005698.4',\n",
       " 'NM_001294338.2',\n",
       " 'NM_000298.6',\n",
       " 'NM_018489.3',\n",
       " 'NM_139119.3',\n",
       " 'NM_001282860.2',\n",
       " 'NM_006912.6',\n",
       " 'NM_014949.4',\n",
       " 'NM_001162383.2',\n",
       " 'NM_003145.4',\n",
       " 'NM_020131.5',\n",
       " 'NM_001093725.2',\n",
       " 'NM_198406.3',\n",
       " 'NM_015327.3',\n",
       " 'NM_144580.3',\n",
       " 'NM_001004319.3',\n",
       " 'NM_005998.5',\n",
       " 'NM_005920.4',\n",
       " 'NM_178229.5',\n",
       " 'NM_006617.2',\n",
       " 'NM_001878.4',\n",
       " 'NM_001370150.2',\n",
       " 'NM_145729.3',\n",
       " 'NM_004494.3',\n",
       " 'NM_003975.4',\n",
       " 'NM_014215.3',\n",
       " 'NM_198236.3',\n",
       " 'NM_001004341.2',\n",
       " 'NM_001145312.3',\n",
       " 'NM_031281.3',\n",
       " 'NM_031282.3',\n",
       " 'NM_052939.4',\n",
       " 'NM_030764.4',\n",
       " 'NM_052938.5',\n",
       " 'NM_005894.3',\n",
       " 'NM_001395415.1',\n",
       " 'NM_001764.3',\n",
       " 'NM_001004475.1',\n",
       " 'NM_001004476.2',\n",
       " 'NM_001005189.2',\n",
       " 'NM_001160325.2',\n",
       " 'NM_001004477.1',\n",
       " 'NM_003126.4',\n",
       " 'NM_001005279.3',\n",
       " 'NM_001005327.3',\n",
       " 'NM_001005185.2',\n",
       " 'NM_001005278.2',\n",
       " 'NM_001320010.2',\n",
       " 'NM_004833.3',\n",
       " 'NM_001004469.1',\n",
       " 'NM_000567.3',\n",
       " 'NM_001013661.1',\n",
       " 'NM_012337.3',\n",
       " 'NM_003564.3',\n",
       " 'NM_001135050.2',\n",
       " 'NM_033438.4',\n",
       " 'NM_145167.3',\n",
       " 'NM_002241.5',\n",
       " 'NM_052868.6',\n",
       " 'NM_015726.4',\n",
       " 'NM_002857.4',\n",
       " 'NM_004371.4',\n",
       " 'NM_001184714.2',\n",
       " 'NM_003874.4',\n",
       " 'NM_003037.5',\n",
       " 'NM_001778.4',\n",
       " 'NM_016382.4',\n",
       " 'NM_017625.3',\n",
       " 'NM_080878.3',\n",
       " 'NM_016946.6',\n",
       " 'NM_001113207.2',\n",
       " 'NM_007122.5',\n",
       " 'NM_001025598.2',\n",
       " 'NM_030916.3',\n",
       " 'NM_012394.4',\n",
       " 'NM_032998.3',\n",
       " 'NM_003779.4',\n",
       " 'NM_005099.6',\n",
       " 'NM_001643.2',\n",
       " 'NM_005122.5',\n",
       " 'NM_000530.8',\n",
       " 'NM_001013625.4',\n",
       " 'NM_000569.8',\n",
       " 'NM_001244753.2',\n",
       " 'NM_015441.3',\n",
       " 'NM_182581.4',\n",
       " 'NM_053282.5',\n",
       " 'NM_001394065.1',\n",
       " 'NM_003617.4',\n",
       " 'NM_177398.4',\n",
       " 'NM_006917.5',\n",
       " 'NM_000696.4',\n",
       " 'NM_019026.6',\n",
       " 'NM_001017961.5',\n",
       " 'NM_053053.4',\n",
       " 'NM_199351.3',\n",
       " 'NM_005814.3',\n",
       " 'NM_198053.3',\n",
       " 'NM_003851.3',\n",
       " 'NM_018417.6',\n",
       " 'NM_001143674.4',\n",
       " 'NM_001375883.1',\n",
       " 'NM_003175.4',\n",
       " 'NM_001937.5',\n",
       " 'NM_013330.5',\n",
       " 'NM_001300969.2',\n",
       " 'NM_006996.3',\n",
       " 'NM_000130.5',\n",
       " 'NM_003005.4',\n",
       " 'NM_000655.5',\n",
       " 'NM_000450.2',\n",
       " 'NM_033418.4',\n",
       " 'NM_020423.7',\n",
       " 'NM_014970.4',\n",
       " 'NM_000261.2',\n",
       " 'NM_003762.5',\n",
       " 'NM_153747.2',\n",
       " 'NM_005092.4',\n",
       " 'NM_003326.5',\n",
       " 'NM_178527.4',\n",
       " 'NM_198493.3',\n",
       " 'NM_001387287.1',\n",
       " 'NM_000488.4',\n",
       " 'NM_172071.4',\n",
       " 'NM_022100.3',\n",
       " 'NM_014656.3',\n",
       " 'NM_003285.3',\n",
       " 'NM_022457.7',\n",
       " 'NM_004319.3',\n",
       " 'NM_033127.4',\n",
       " 'NM_001395331.1',\n",
       " 'NM_004673.4',\n",
       " 'NM_007314.4',\n",
       " 'NM_014625.4',\n",
       " 'NM_001199260.2',\n",
       " 'NM_032360.4',\n",
       " 'NM_005819.6',\n",
       " 'NM_001009992.1',\n",
       " 'NM_001033044.4',\n",
       " 'NM_172000.4',\n",
       " 'NM_021133.4',\n",
       " 'NM_002928.4',\n",
       " 'NM_001102450.3',\n",
       " 'NM_030933.4',\n",
       " 'NM_015039.4',\n",
       " 'NM_000433.4',\n",
       " 'NM_005717.4',\n",
       " 'NM_203454.3',\n",
       " 'NM_015101.4',\n",
       " 'NM_025191.4',\n",
       " 'NM_052966.4',\n",
       " 'NM_030934.5',\n",
       " 'NM_006469.5',\n",
       " 'NM_003292.3',\n",
       " 'NM_002597.5',\n",
       " 'NM_000963.4',\n",
       " 'NM_199051.3',\n",
       " 'NM_001199261.3',\n",
       " 'NM_197962.3',\n",
       " 'NM_003783.3',\n",
       " 'NM_198503.5',\n",
       " 'NM_001994.3',\n",
       " 'NM_018136.5',\n",
       " 'NM_194314.3',\n",
       " 'NM_001195215.2',\n",
       " 'NM_001376861.1',\n",
       " 'NM_001281293.2',\n",
       " 'NM_014875.3',\n",
       " 'NM_001031725.6',\n",
       " 'NM_001252102.2',\n",
       " 'NM_000069.3',\n",
       " 'NM_001270601.2',\n",
       " 'NM_001288565.2',\n",
       " 'NM_001276345.2',\n",
       " 'NM_005558.4',\n",
       " 'NM_003281.4',\n",
       " 'NM_012396.5',\n",
       " 'NM_004078.3',\n",
       " 'NM_012134.3',\n",
       " 'NM_138795.4',\n",
       " 'NM_002832.4',\n",
       " 'NM_014176.4',\n",
       " 'NM_177402.5',\n",
       " 'NM_006618.5',\n",
       " 'NM_002871.5',\n",
       " 'NM_021633.4',\n",
       " 'NM_015999.6',\n",
       " 'NM_016243.3',\n",
       " 'NM_002479.6',\n",
       " 'NM_004997.3',\n",
       " 'NM_001276.4',\n",
       " 'NM_003465.3',\n",
       " 'NM_002023.5',\n",
       " 'NM_018208.4',\n",
       " 'NM_000537.4',\n",
       " 'NM_002256.4',\n",
       " 'NM_198447.2',\n",
       " 'NM_014935.5',\n",
       " 'NM_032833.5',\n",
       " 'NM_001377334.1',\n",
       " 'NM_201630.2',\n",
       " 'NM_203376.2',\n",
       " 'NM_005057.4',\n",
       " 'NM_015375.3',\n",
       " 'NM_030952.3',\n",
       " 'NM_018203.3',\n",
       " 'NM_001199050.2',\n",
       " 'NM_001397426.2',\n",
       " 'NM_001973.4',\n",
       " 'NM_033102.3',\n",
       " 'NM_022731.5',\n",
       " 'NM_003929.3',\n",
       " 'NM_173854.6',\n",
       " 'NM_152491.5',\n",
       " 'NM_052934.4',\n",
       " 'NM_001164522.3',\n",
       " 'NM_001910.4',\n",
       " 'NM_000707.5',\n",
       " 'NM_001123168.3',\n",
       " 'NM_006893.3',\n",
       " 'NM_000572.3',\n",
       " 'NM_005449.5',\n",
       " 'NM_002644.4',\n",
       " 'NM_001170631.2',\n",
       " 'NM_023938.6',\n",
       " 'NM_018566.4',\n",
       " 'NM_001025109.2',\n",
       " 'NM_025179.4',\n",
       " 'NM_000228.3',\n",
       " 'NM_152485.4',\n",
       " 'NM_006147.4',\n",
       " 'NM_172362.3',\n",
       " 'NM_001164688.2',\n",
       " 'NM_021194.3',\n",
       " 'NM_002497.4',\n",
       " 'NM_014873.3',\n",
       " 'NM_015434.4',\n",
       " 'NM_018252.3',\n",
       " 'NM_018664.3',\n",
       " 'NM_015471.4',\n",
       " 'NM_001024601.3',\n",
       " 'NM_144567.5',\n",
       " 'NM_005401.5',\n",
       " 'NM_206933.4',\n",
       " 'NM_001438.4',\n",
       " 'NM_018040.5',\n",
       " 'NM_001355457.3',\n",
       " 'NM_018713.3',\n",
       " 'NM_004446.3',\n",
       " 'NM_006085.6',\n",
       " 'NM_012414.4',\n",
       " 'NM_007207.6',\n",
       " 'NM_024746.4',\n",
       " 'NM_005681.4',\n",
       " 'NM_022831.4',\n",
       " 'NM_003268.6',\n",
       " 'NM_017982.4',\n",
       " 'NM_001143962.2',\n",
       " 'NM_001031685.3',\n",
       " 'NM_002533.4',\n",
       " 'NM_001379403.1',\n",
       " 'NM_002296.4',\n",
       " 'NM_018212.6',\n",
       " 'NM_014698.3',\n",
       " 'NM_020997.4',\n",
       " 'NM_013328.4',\n",
       " 'NM_003240.5',\n",
       " 'NM_152608.4',\n",
       " 'NM_022735.4',\n",
       " 'NM_001366245.2',\n",
       " 'NM_001618.4',\n",
       " 'NM_002221.4',\n",
       " 'NM_001394014.1',\n",
       " 'NM_023007.3',\n",
       " 'NM_003395.4',\n",
       " 'NM_024319.4',\n",
       " 'NM_181463.3',\n",
       " 'NM_145214.3',\n",
       " 'NM_016102.4',\n",
       " 'NM_003493.3',\n",
       " 'NM_033445.3',\n",
       " 'NM_145257.5',\n",
       " 'NM_001100.4',\n",
       " 'NM_018230.3',\n",
       " 'NM_012089.3',\n",
       " 'NM_014409.4',\n",
       " 'NM_001258311.2',\n",
       " 'NM_001384479.1',\n",
       " 'NM_032800.3',\n",
       " 'NM_024525.5',\n",
       " 'NM_198552.3',\n",
       " 'NM_152379.4',\n",
       " 'NM_175876.5',\n",
       " 'NM_022051.3',\n",
       " 'NM_020808.5',\n",
       " 'NM_014801.4',\n",
       " 'NM_005646.4',\n",
       " 'NM_182972.3',\n",
       " 'NM_014765.3',\n",
       " 'NM_015014.4',\n",
       " 'NM_016374.6',\n",
       " 'NM_152490.5',\n",
       " 'NM_001098722.2',\n",
       " 'NM_000081.4',\n",
       " 'NM_002508.3',\n",
       " 'NM_019891.4',\n",
       " 'NM_018072.6',\n",
       " 'NM_001276687.2',\n",
       " 'NM_021186.5',\n",
       " 'NM_022469.4',\n",
       " 'NM_001364886.1',\n",
       " 'NM_000143.4',\n",
       " 'NM_014322.3',\n",
       " 'NM_001381853.1',\n",
       " 'NM_001004343.3',\n",
       " 'NM_001372062.1',\n",
       " 'NM_014812.3',\n",
       " 'NM_005465.7',\n",
       " 'NM_001126.5',\n",
       " 'NM_001395959.1',\n",
       " 'NM_031844.3',\n",
       " 'NM_001167740.2',\n",
       " 'NM_022366.3',\n",
       " 'NM_001323342.2',\n",
       " 'NM_020394.5',\n",
       " 'NM_033213.5',\n",
       " 'NM_001142572.2',\n",
       " 'NM_001297568.2',\n",
       " 'NM_032752.3',\n",
       " 'NM_001004492.2',\n",
       " 'NM_198074.6',\n",
       " 'NM_001005487.2',\n",
       " 'NM_001005286.2',\n",
       " 'NM_001355292.3',\n",
       " 'NM_012353.3',\n",
       " 'NM_001001966.2',\n",
       " 'NM_001001959.1',\n",
       " 'NM_001004695.2',\n",
       " 'NM_001004692.2',\n",
       " 'NM_001004691.1',\n",
       " 'NM_001004694.3',\n",
       " 'NM_001001821.1',\n",
       " 'NM_001004693.2',\n",
       " 'NM_001001964.2',\n",
       " 'NM_001001827.2',\n",
       " 'NM_001001824.2',\n",
       " 'NM_001004734.4',\n",
       " 'NM_001085474.2',\n",
       " 'NM_030645.3',\n",
       " 'NM_017865.4',\n",
       " 'NM_001077710.3',\n",
       " 'NM_001002919.3',\n",
       " 'NM_152834.4',\n",
       " 'NM_012293.3',\n",
       " 'NM_001303052.2',\n",
       " 'NM_003310.5',\n",
       " 'NM_018269.4',\n",
       " 'NM_002936.6',\n",
       " 'NM_207315.4',\n",
       " 'NM_020738.4',\n",
       " 'NM_138799.4',\n",
       " 'NM_004763.5',\n",
       " 'NM_003183.6',\n",
       " 'NM_006826.4',\n",
       " 'NM_001037160.3',\n",
       " 'NM_002539.3',\n",
       " 'NM_024894.4',\n",
       " ...]"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start_in_middle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "0eb40fff-61db-402f-9ef3-f9d7b1b1a0be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>450743</td>\n",
       "      <td>451678</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"OR4F29\"; transcript_id \"NM_001005221....</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>NM_001005221.2</td>\n",
       "      <td>729759</td>\n",
       "      <td>936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>451676</td>\n",
       "      <td>451678</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"OR4F29\"; transcript_id \"NM_001005221....</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>NM_001005221.2</td>\n",
       "      <td>729759</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>450740</td>\n",
       "      <td>450742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"OR4F29\"; transcript_id \"NM_001005221....</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>NM_001005221.2</td>\n",
       "      <td>729759</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>685719</td>\n",
       "      <td>686654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"OR4F16\"; transcript_id \"NM_001005277....</td>\n",
       "      <td>OR4F16</td>\n",
       "      <td>NM_001005277.1</td>\n",
       "      <td>81399</td>\n",
       "      <td>936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>686652</td>\n",
       "      <td>686654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"OR4F16\"; transcript_id \"NM_001005277....</td>\n",
       "      <td>OR4F16</td>\n",
       "      <td>NM_001005277.1</td>\n",
       "      <td>81399</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>685716</td>\n",
       "      <td>685718</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"OR4F16\"; transcript_id \"NM_001005277....</td>\n",
       "      <td>OR4F16</td>\n",
       "      <td>NM_001005277.1</td>\n",
       "      <td>81399</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>959215</td>\n",
       "      <td>959240</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658.4</td>\n",
       "      <td>26155</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>958929</td>\n",
       "      <td>959081</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658.4</td>\n",
       "      <td>26155</td>\n",
       "      <td>153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>957099</td>\n",
       "      <td>957273</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658.4</td>\n",
       "      <td>26155</td>\n",
       "      <td>175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>956894</td>\n",
       "      <td>957025</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658.4</td>\n",
       "      <td>26155</td>\n",
       "      <td>132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>956095</td>\n",
       "      <td>956215</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658.4</td>\n",
       "      <td>26155</td>\n",
       "      <td>121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>955923</td>\n",
       "      <td>956013</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658.4</td>\n",
       "      <td>26155</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>954004</td>\n",
       "      <td>954082</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658.4</td>\n",
       "      <td>26155</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>953782</td>\n",
       "      <td>953892</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658.4</td>\n",
       "      <td>26155</td>\n",
       "      <td>111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>953175</td>\n",
       "      <td>953288</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658.4</td>\n",
       "      <td>26155</td>\n",
       "      <td>114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>952412</td>\n",
       "      <td>952600</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658.4</td>\n",
       "      <td>26155</td>\n",
       "      <td>189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>952000</td>\n",
       "      <td>952139</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658.4</td>\n",
       "      <td>26155</td>\n",
       "      <td>140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>951127</td>\n",
       "      <td>951238</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658.4</td>\n",
       "      <td>26155</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>948490</td>\n",
       "      <td>948603</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658.4</td>\n",
       "      <td>26155</td>\n",
       "      <td>114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>948131</td>\n",
       "      <td>948232</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658.4</td>\n",
       "      <td>26155</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>946402</td>\n",
       "      <td>946545</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658.4</td>\n",
       "      <td>26155</td>\n",
       "      <td>144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>946173</td>\n",
       "      <td>946286</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658.4</td>\n",
       "      <td>26155</td>\n",
       "      <td>114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>945518</td>\n",
       "      <td>945653</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658.4</td>\n",
       "      <td>26155</td>\n",
       "      <td>136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>945057</td>\n",
       "      <td>945146</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658.4</td>\n",
       "      <td>26155</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>944697</td>\n",
       "      <td>944800</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658.4</td>\n",
       "      <td>26155</td>\n",
       "      <td>104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>959238</td>\n",
       "      <td>959240</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658.4</td>\n",
       "      <td>26155</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>944694</td>\n",
       "      <td>944696</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658.4</td>\n",
       "      <td>26155</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>978881</td>\n",
       "      <td>981029</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"PERM1\"; transcript_id \"NM_001394713.1...</td>\n",
       "      <td>PERM1</td>\n",
       "      <td>NM_001394713.1</td>\n",
       "      <td>84808</td>\n",
       "      <td>2149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>976499</td>\n",
       "      <td>976624</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"PERM1\"; transcript_id \"NM_001394713.1...</td>\n",
       "      <td>PERM1</td>\n",
       "      <td>NM_001394713.1</td>\n",
       "      <td>84808</td>\n",
       "      <td>126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>976175</td>\n",
       "      <td>976269</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"PERM1\"; transcript_id \"NM_001394713.1...</td>\n",
       "      <td>PERM1</td>\n",
       "      <td>NM_001394713.1</td>\n",
       "      <td>84808</td>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>981027</td>\n",
       "      <td>981029</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"PERM1\"; transcript_id \"NM_001394713.1...</td>\n",
       "      <td>PERM1</td>\n",
       "      <td>NM_001394713.1</td>\n",
       "      <td>84808</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>976172</td>\n",
       "      <td>976174</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"PERM1\"; transcript_id \"NM_001394713.1...</td>\n",
       "      <td>PERM1</td>\n",
       "      <td>NM_001394713.1</td>\n",
       "      <td>84808</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>999866</td>\n",
       "      <td>999973</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"HES4\"; transcript_id \"NM_021170.4\"; d...</td>\n",
       "      <td>HES4</td>\n",
       "      <td>NM_021170.4</td>\n",
       "      <td>57801</td>\n",
       "      <td>108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>999692</td>\n",
       "      <td>999787</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"HES4\"; transcript_id \"NM_021170.4\"; d...</td>\n",
       "      <td>HES4</td>\n",
       "      <td>NM_021170.4</td>\n",
       "      <td>57801</td>\n",
       "      <td>96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>999526</td>\n",
       "      <td>999613</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"HES4\"; transcript_id \"NM_021170.4\"; d...</td>\n",
       "      <td>HES4</td>\n",
       "      <td>NM_021170.4</td>\n",
       "      <td>57801</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>999062</td>\n",
       "      <td>999432</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"HES4\"; transcript_id \"NM_021170.4\"; d...</td>\n",
       "      <td>HES4</td>\n",
       "      <td>NM_021170.4</td>\n",
       "      <td>57801</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>999971</td>\n",
       "      <td>999973</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"HES4\"; transcript_id \"NM_021170.4\"; d...</td>\n",
       "      <td>HES4</td>\n",
       "      <td>NM_021170.4</td>\n",
       "      <td>57801</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>999059</td>\n",
       "      <td>999061</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"HES4\"; transcript_id \"NM_021170.4\"; d...</td>\n",
       "      <td>HES4</td>\n",
       "      <td>NM_021170.4</td>\n",
       "      <td>57801</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1071820</td>\n",
       "      <td>1072566</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"RNF223\"; transcript_id \"NM_001205252....</td>\n",
       "      <td>RNF223</td>\n",
       "      <td>NM_001205252.2</td>\n",
       "      <td>401934</td>\n",
       "      <td>747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1072564</td>\n",
       "      <td>1072566</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"RNF223\"; transcript_id \"NM_001205252....</td>\n",
       "      <td>RNF223</td>\n",
       "      <td>NM_001205252.2</td>\n",
       "      <td>401934</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>1071817</td>\n",
       "      <td>1071819</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"RNF223\"; transcript_id \"NM_001205252....</td>\n",
       "      <td>RNF223</td>\n",
       "      <td>NM_001205252.2</td>\n",
       "      <td>401934</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1091472</td>\n",
       "      <td>1091543</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"C1orf159\"; transcript_id \"NM_017891.5...</td>\n",
       "      <td>C1orf159</td>\n",
       "      <td>NM_017891.5</td>\n",
       "      <td>54991</td>\n",
       "      <td>72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1090353</td>\n",
       "      <td>1090428</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"C1orf159\"; transcript_id \"NM_017891.5...</td>\n",
       "      <td>C1orf159</td>\n",
       "      <td>NM_017891.5</td>\n",
       "      <td>54991</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1087502</td>\n",
       "      <td>1087597</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"C1orf159\"; transcript_id \"NM_017891.5...</td>\n",
       "      <td>C1orf159</td>\n",
       "      <td>NM_017891.5</td>\n",
       "      <td>54991</td>\n",
       "      <td>96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1087139</td>\n",
       "      <td>1087204</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"C1orf159\"; transcript_id \"NM_017891.5...</td>\n",
       "      <td>C1orf159</td>\n",
       "      <td>NM_017891.5</td>\n",
       "      <td>54991</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1085878</td>\n",
       "      <td>1086012</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"C1orf159\"; transcript_id \"NM_017891.5...</td>\n",
       "      <td>C1orf159</td>\n",
       "      <td>NM_017891.5</td>\n",
       "      <td>54991</td>\n",
       "      <td>135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1084481</td>\n",
       "      <td>1084506</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"C1orf159\"; transcript_id \"NM_017891.5...</td>\n",
       "      <td>C1orf159</td>\n",
       "      <td>NM_017891.5</td>\n",
       "      <td>54991</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1084353</td>\n",
       "      <td>1084383</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"C1orf159\"; transcript_id \"NM_017891.5...</td>\n",
       "      <td>C1orf159</td>\n",
       "      <td>NM_017891.5</td>\n",
       "      <td>54991</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1082896</td>\n",
       "      <td>1082987</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"C1orf159\"; transcript_id \"NM_017891.5...</td>\n",
       "      <td>C1orf159</td>\n",
       "      <td>NM_017891.5</td>\n",
       "      <td>54991</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1091541</td>\n",
       "      <td>1091543</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"C1orf159\"; transcript_id \"NM_017891.5...</td>\n",
       "      <td>C1orf159</td>\n",
       "      <td>NM_017891.5</td>\n",
       "      <td>54991</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqname      source      feature    start      end score strand  frame  \\\n",
       "4      chr1  BestRefSeq          CDS   450743   451678     .      -      0   \n",
       "5      chr1  BestRefSeq  start_codon   451676   451678     .      -      0   \n",
       "6      chr1  BestRefSeq   stop_codon   450740   450742     .      -      0   \n",
       "7      chr1  BestRefSeq          CDS   685719   686654     .      -      0   \n",
       "8      chr1  BestRefSeq  start_codon   686652   686654     .      -      0   \n",
       "9      chr1  BestRefSeq   stop_codon   685716   685718     .      -      0   \n",
       "26     chr1  BestRefSeq          CDS   959215   959240     .      -      0   \n",
       "27     chr1  BestRefSeq          CDS   958929   959081     .      -      1   \n",
       "28     chr1  BestRefSeq          CDS   957099   957273     .      -      1   \n",
       "29     chr1  BestRefSeq          CDS   956894   957025     .      -      0   \n",
       "30     chr1  BestRefSeq          CDS   956095   956215     .      -      0   \n",
       "31     chr1  BestRefSeq          CDS   955923   956013     .      -      2   \n",
       "32     chr1  BestRefSeq          CDS   954004   954082     .      -      1   \n",
       "33     chr1  BestRefSeq          CDS   953782   953892     .      -      0   \n",
       "34     chr1  BestRefSeq          CDS   953175   953288     .      -      0   \n",
       "35     chr1  BestRefSeq          CDS   952412   952600     .      -      0   \n",
       "36     chr1  BestRefSeq          CDS   952000   952139     .      -      0   \n",
       "37     chr1  BestRefSeq          CDS   951127   951238     .      -      1   \n",
       "38     chr1  BestRefSeq          CDS   948490   948603     .      -      0   \n",
       "39     chr1  BestRefSeq          CDS   948131   948232     .      -      0   \n",
       "40     chr1  BestRefSeq          CDS   946402   946545     .      -      0   \n",
       "41     chr1  BestRefSeq          CDS   946173   946286     .      -      0   \n",
       "42     chr1  BestRefSeq          CDS   945518   945653     .      -      0   \n",
       "43     chr1  BestRefSeq          CDS   945057   945146     .      -      2   \n",
       "44     chr1  BestRefSeq          CDS   944697   944800     .      -      2   \n",
       "45     chr1  BestRefSeq  start_codon   959238   959240     .      -      0   \n",
       "46     chr1  BestRefSeq   stop_codon   944694   944696     .      -      0   \n",
       "79     chr1  BestRefSeq          CDS   978881   981029     .      -      0   \n",
       "80     chr1  BestRefSeq          CDS   976499   976624     .      -      2   \n",
       "81     chr1  BestRefSeq          CDS   976175   976269     .      -      2   \n",
       "82     chr1  BestRefSeq  start_codon   981027   981029     .      -      0   \n",
       "83     chr1  BestRefSeq   stop_codon   976172   976174     .      -      0   \n",
       "84     chr1  BestRefSeq          CDS   999866   999973     .      -      0   \n",
       "85     chr1  BestRefSeq          CDS   999692   999787     .      -      0   \n",
       "86     chr1  BestRefSeq          CDS   999526   999613     .      -      0   \n",
       "87     chr1  BestRefSeq          CDS   999062   999432     .      -      2   \n",
       "88     chr1  BestRefSeq  start_codon   999971   999973     .      -      0   \n",
       "89     chr1  BestRefSeq   stop_codon   999059   999061     .      -      0   \n",
       "132    chr1  BestRefSeq          CDS  1071820  1072566     .      -      0   \n",
       "133    chr1  BestRefSeq  start_codon  1072564  1072566     .      -      0   \n",
       "134    chr1  BestRefSeq   stop_codon  1071817  1071819     .      -      0   \n",
       "135    chr1  BestRefSeq          CDS  1091472  1091543     .      -      0   \n",
       "136    chr1  BestRefSeq          CDS  1090353  1090428     .      -      0   \n",
       "137    chr1  BestRefSeq          CDS  1087502  1087597     .      -      2   \n",
       "138    chr1  BestRefSeq          CDS  1087139  1087204     .      -      2   \n",
       "139    chr1  BestRefSeq          CDS  1085878  1086012     .      -      2   \n",
       "140    chr1  BestRefSeq          CDS  1084481  1084506     .      -      2   \n",
       "141    chr1  BestRefSeq          CDS  1084353  1084383     .      -      0   \n",
       "142    chr1  BestRefSeq          CDS  1082896  1082987     .      -      2   \n",
       "143    chr1  BestRefSeq  start_codon  1091541  1091543     .      -      0   \n",
       "\n",
       "                                            attributes   gene_id  \\\n",
       "4    gene_id \"OR4F29\"; transcript_id \"NM_001005221....    OR4F29   \n",
       "5    gene_id \"OR4F29\"; transcript_id \"NM_001005221....    OR4F29   \n",
       "6    gene_id \"OR4F29\"; transcript_id \"NM_001005221....    OR4F29   \n",
       "7    gene_id \"OR4F16\"; transcript_id \"NM_001005277....    OR4F16   \n",
       "8    gene_id \"OR4F16\"; transcript_id \"NM_001005277....    OR4F16   \n",
       "9    gene_id \"OR4F16\"; transcript_id \"NM_001005277....    OR4F16   \n",
       "26   gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...     NOC2L   \n",
       "27   gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...     NOC2L   \n",
       "28   gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...     NOC2L   \n",
       "29   gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...     NOC2L   \n",
       "30   gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...     NOC2L   \n",
       "31   gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...     NOC2L   \n",
       "32   gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...     NOC2L   \n",
       "33   gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...     NOC2L   \n",
       "34   gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...     NOC2L   \n",
       "35   gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...     NOC2L   \n",
       "36   gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...     NOC2L   \n",
       "37   gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...     NOC2L   \n",
       "38   gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...     NOC2L   \n",
       "39   gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...     NOC2L   \n",
       "40   gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...     NOC2L   \n",
       "41   gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...     NOC2L   \n",
       "42   gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...     NOC2L   \n",
       "43   gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...     NOC2L   \n",
       "44   gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...     NOC2L   \n",
       "45   gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...     NOC2L   \n",
       "46   gene_id \"NOC2L\"; transcript_id \"NM_015658.4\"; ...     NOC2L   \n",
       "79   gene_id \"PERM1\"; transcript_id \"NM_001394713.1...     PERM1   \n",
       "80   gene_id \"PERM1\"; transcript_id \"NM_001394713.1...     PERM1   \n",
       "81   gene_id \"PERM1\"; transcript_id \"NM_001394713.1...     PERM1   \n",
       "82   gene_id \"PERM1\"; transcript_id \"NM_001394713.1...     PERM1   \n",
       "83   gene_id \"PERM1\"; transcript_id \"NM_001394713.1...     PERM1   \n",
       "84   gene_id \"HES4\"; transcript_id \"NM_021170.4\"; d...      HES4   \n",
       "85   gene_id \"HES4\"; transcript_id \"NM_021170.4\"; d...      HES4   \n",
       "86   gene_id \"HES4\"; transcript_id \"NM_021170.4\"; d...      HES4   \n",
       "87   gene_id \"HES4\"; transcript_id \"NM_021170.4\"; d...      HES4   \n",
       "88   gene_id \"HES4\"; transcript_id \"NM_021170.4\"; d...      HES4   \n",
       "89   gene_id \"HES4\"; transcript_id \"NM_021170.4\"; d...      HES4   \n",
       "132  gene_id \"RNF223\"; transcript_id \"NM_001205252....    RNF223   \n",
       "133  gene_id \"RNF223\"; transcript_id \"NM_001205252....    RNF223   \n",
       "134  gene_id \"RNF223\"; transcript_id \"NM_001205252....    RNF223   \n",
       "135  gene_id \"C1orf159\"; transcript_id \"NM_017891.5...  C1orf159   \n",
       "136  gene_id \"C1orf159\"; transcript_id \"NM_017891.5...  C1orf159   \n",
       "137  gene_id \"C1orf159\"; transcript_id \"NM_017891.5...  C1orf159   \n",
       "138  gene_id \"C1orf159\"; transcript_id \"NM_017891.5...  C1orf159   \n",
       "139  gene_id \"C1orf159\"; transcript_id \"NM_017891.5...  C1orf159   \n",
       "140  gene_id \"C1orf159\"; transcript_id \"NM_017891.5...  C1orf159   \n",
       "141  gene_id \"C1orf159\"; transcript_id \"NM_017891.5...  C1orf159   \n",
       "142  gene_id \"C1orf159\"; transcript_id \"NM_017891.5...  C1orf159   \n",
       "143  gene_id \"C1orf159\"; transcript_id \"NM_017891.5...  C1orf159   \n",
       "\n",
       "      transcript_id  entrez_id  length  \n",
       "4    NM_001005221.2     729759     936  \n",
       "5    NM_001005221.2     729759       3  \n",
       "6    NM_001005221.2     729759       3  \n",
       "7    NM_001005277.1      81399     936  \n",
       "8    NM_001005277.1      81399       3  \n",
       "9    NM_001005277.1      81399       3  \n",
       "26      NM_015658.4      26155      26  \n",
       "27      NM_015658.4      26155     153  \n",
       "28      NM_015658.4      26155     175  \n",
       "29      NM_015658.4      26155     132  \n",
       "30      NM_015658.4      26155     121  \n",
       "31      NM_015658.4      26155      91  \n",
       "32      NM_015658.4      26155      79  \n",
       "33      NM_015658.4      26155     111  \n",
       "34      NM_015658.4      26155     114  \n",
       "35      NM_015658.4      26155     189  \n",
       "36      NM_015658.4      26155     140  \n",
       "37      NM_015658.4      26155     112  \n",
       "38      NM_015658.4      26155     114  \n",
       "39      NM_015658.4      26155     102  \n",
       "40      NM_015658.4      26155     144  \n",
       "41      NM_015658.4      26155     114  \n",
       "42      NM_015658.4      26155     136  \n",
       "43      NM_015658.4      26155      90  \n",
       "44      NM_015658.4      26155     104  \n",
       "45      NM_015658.4      26155       3  \n",
       "46      NM_015658.4      26155       3  \n",
       "79   NM_001394713.1      84808    2149  \n",
       "80   NM_001394713.1      84808     126  \n",
       "81   NM_001394713.1      84808      95  \n",
       "82   NM_001394713.1      84808       3  \n",
       "83   NM_001394713.1      84808       3  \n",
       "84      NM_021170.4      57801     108  \n",
       "85      NM_021170.4      57801      96  \n",
       "86      NM_021170.4      57801      88  \n",
       "87      NM_021170.4      57801     371  \n",
       "88      NM_021170.4      57801       3  \n",
       "89      NM_021170.4      57801       3  \n",
       "132  NM_001205252.2     401934     747  \n",
       "133  NM_001205252.2     401934       3  \n",
       "134  NM_001205252.2     401934       3  \n",
       "135     NM_017891.5      54991      72  \n",
       "136     NM_017891.5      54991      76  \n",
       "137     NM_017891.5      54991      96  \n",
       "138     NM_017891.5      54991      66  \n",
       "139     NM_017891.5      54991     135  \n",
       "140     NM_017891.5      54991      26  \n",
       "141     NM_017891.5      54991      31  \n",
       "142     NM_017891.5      54991      92  \n",
       "143     NM_017891.5      54991       3  "
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_cds[gtf_cds['strand'] == '-'].head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b96b6da-8460-482d-8723-d9ea379f7580",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "target-genes",
   "language": "python",
   "name": "target-genes"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
